---
title: "Codon Usage"
author: "Abhishek Jain"
date: "18 June 2019"
output: 
  md_document:
    variant: markdown_github
editor_options:
  chunk_output_type: console
---

## Data exploration

```{r setup}

if(!require(tidyverse)){
    install.packages("tidyverse")
    library(tidyverse)
}
if(!require(glmnet)){
    install.packages("glmnet")
    library(glmnet)
}
if(!require(ggplot2)){
    install.packages("ggplot2")
    library(tidyverse)
}

```

```{R Load Data}

ref_raw <- read_rds("data/Sun_mutation_UTRs.rds")
  #Get sequences from ref_raw in a separate vector
  cds_seq <- ref_raw$CDS_seq
  
sc_raw <-read_tsv("data/YPD_scRNA.tab")

#Load Manually created motifs list into a vector
#motifs_raw <- scan("data/list_motifs.txt", character())
list_genes <- scan("data/genes_terminators_list.txt", character())

#motifs_cheng = c("TGTAAATA", "TGCAT", "TTTTTTA", "ATATTC")

#this is new karsten/Chan et al 2018 data
  dr_raw <- read_tsv("data/new_karsten_dr_data.txt")
  codons_raw <- scan("data/codons.txt", character())
  
```

```{R clean data}
#create a dataframe with codon usage for each gene
ref_codons_freq <- ref_raw %>%
  select(genename, CDS_seq, TTT:GGG) %>%
    rename(geneName = genename)

#Calculate proportional codon usage for each transcript & remove one codon
ref_codons <- ref_codons_freq %>%
  mutate_at(vars(TTT:GGG), funs(./nchar(CDS_seq))) %>%
    select(-TTT, -CDS_seq) 
  

#change colnames, add means and filter missing values 
dr_raw2 <- dr_raw %>%
  rename_all(~c("geneName","gene","hlife_r1","hlife_r2")) %>%
    dplyr::select(-gene) %>%
      mutate(hlife = rowMeans(cbind(hlife_r1, hlife_r2), na.rm = TRUE)) %>% 
        filter(is.finite(hlife)) 

#cleaner df
dr_data <- dr_raw2 %>%
          dplyr::select(geneName, hlife)
  
```

Add codon usage to each dataset

```{R Add Codon Information}

#add dr data for Chan et al
ref_codons_dr <- left_join(ref_codons, dr_data, by = "geneName") %>%
  filter(is.finite(hlife))

```

```{R Multiple Linear Modelling}

#make model data
model_data <- ref_codons_dr %>%
  #convert to hlife to log2fold
    mutate(log2_hlife = log2(hlife)) %>%
      select(-hlife, -CDS_seq)


#selected_data
selected_data <- model_data %>%
  filter(geneName %in% list_genes)

#Create a subset of model_data for linear modelling

  #creates an operator for Not %in%
  '%!in%' <- function(x,y)!('%in%'(x,y))
  
    model_data_sub <- model_data %>%
      filter(geneName %!in% list_genes)

#list of codons in model data
codons <- colnames(subset(model_data, 
                          select = -c(geneName, log2_hlife) ))

#full multiple linear model
lm_codons_dr <- lm(data = model_combined, 
                     paste("log2_hlife", 
                           paste(codons, 
                                 collapse = " + "),
                           sep = " ~ "))



#multiple linear model of subset
lm_codons_sub <- lm(data = model_data_sub, 
                     paste("log2_hlife", 
                           paste(codons, 
                                 collapse = " + "),
                           sep = " ~ "))

```

```{R Prediction from LM}

predicted_data <- selected_data %>%
  mutate(pred_log2hlife = predict(lm_codons_sub, selected_data))

predicted_data %>%
  ggplot(aes(x = log2_hlife, y = pred_log2hlife, col = geneName)) +
    geom_point() + 
    scale_x_continuous(limits = c(0, 5)) + 
    scale_y_continuous(limits = c(0, 5)) +
    geom_abline() +
    #geom_smooth(method = "lm", formula = y ~ x) +
#    theme_coding() + 
    labs(y = "Predicted Half life", 
         x = "Measured Half Life")

```


```{R Glmnet Matrix}
#Chan et al
model_matrix <- model_data %>%
    dplyr::select(-geneName, -log2_hlife) %>%
      #sapply(., strtoi)%>%
        data.matrix()

model_matrix_sub <- model_data_sub %>%
    dplyr::select(-geneName, -log2_hlife) %>%
      #sapply(., strtoi)%>%
        data.matrix()

pred_matrix_sel <- selected_data %>%
    dplyr::select(-geneName, -log2_hlife) %>%
      data.matrix()


```

```{Performing Glmnet on Chan et al 2018}

#Cross Validated Glmnet to get lambda.min for extracting coeffs 
cv_glm <- cv.glmnet(x = model_matrix_sub, 
                    y = model_data_sub$log2_hlife,
                    nfolds = 100)

#visualize MSE vs log(lambda plot) for CV glmnet
plot(cv_glm, sub = "DATA: Chan et al. 2018 ")


```


```{R glmnet Linear Modelling}
coefs <- data.matrix(coef(cv_glm, s = "lambda.min"))

#Create a tibble of coefficients ordered by absolute value
sig_glms <- tibble(
      codons_glmnet = rownames(coefs),
      coeff_glmnet = coefs[,1]
    ) %>%
        arrange(desc(abs(coeff_glmnet))) %>%
          filter(codons_glmnet != "(Intercept)") %>%
          filter(coeff_glmnet != 0)

#create multiple lm model for predtion 
lm_codons_glmnet <- lm(data = model_data_sub, 
                     paste("log2_hlife", 
                           paste(sig_glms$codons_glmnet, 
                                 collapse = " + "),
                           sep = " ~ "))

#predict using glmnet
predicted_data_g <- selected_data %>%
  mutate(pred_log2hlife = predict(cv_glm, pred_matrix_sel))

predicted_data_g %>%
  ggplot(aes(x = log2_hlife, y = pred_log2hlife)) +
    geom_point() + 
    scale_x_continuous(limits = c(0, 5)) + 
    scale_y_continuous(limits = c(0, 5)) + 
    geom_smooth(method = "lm", formula = y ~ x) +
#    theme_coding() + 
    labs(y = "Predicted Half life", 
         x = "Measured Half Life")

```

Code for checking if sun et al codon usage values are correct
  (Take may take a couple minutes to run)
  
```{R Alternative Codon Count, eval=FALSE, include=FALSE}
#alternative method (testing if Sun's codon number are correct)
ref_codons2 <- tibble(genename = ref_raw$genename)


new_func <- function(seq_vec, codon_x = "TTC"){
  
  sst <- strsplit(seq_vec, "")
  
  sst2 <- lapply(sst, function(sst){
    do.call(paste0,
            lapply(seq_len(3), 
              function(i) {
                idx <- rep(FALSE, 3);
                idx[i] <- TRUE;
                sst[idx]
              })
            )
    })
  
  #search 
  sapply(sst2, function(ii){
    sum(str_count(ii, codon_x))
  })
    
}


```


Following code for using codon usage model to predict hlife is still incomplete 
```{R Prediction, eval=FALSE, include=FALSE}
predict(lm_motifs_k_dr, selected_data)
predicted_data <- model_data_sel %>%
  mutate(pred_log2hlife = predict(lm_motifs_k_dr, selected_data))

predicted_data %>%
  ggplot(aes(x = log2_hlife, y = pred_log2hlife)) +
    geom_point() + 
    scale_x_continuous(limits = c(0, 4)) + 
    scale_y_continuous(limits = c(0, 4)) + 
    geom_smooth(method = "lm", formula = y ~ x) +
#    theme_coding() + 
    labs(y = "Predicted Half life", 
         x = "Measured Half Life")

```

```{R test combined model}

model_data_codons <- model_data %>%
  select(-TTT, -log2_hlife)

model_matrix_codons <- model_data %>%
    dplyr::select(-geneName, -log2_hlife) %>%
      #sapply(., strtoi)%>%
        data.matrix()

model_matrix_motifs <- model_data_k_dr %>%
    dplyr::select(-geneName, -log2_hlife) %>%
      sapply(., strtoi)%>%
        data.matrix()

model_matrix_combined <- cbind(model_matrix_codons, model_matrix_motifs)


cv_glmnet_com <- cv.glmnet(x = model_matrix_combined, 
                           y = model_data$log2_hlife,
                           nfolds = 100)

#visualize MSE vs log(lambda plot) for CV glmnet
plot(cv_glmnet_com, sub = "DATA: Chan et al. 2018 ")



coefs <- data.matrix(coef(cv_glmnet_com, s = "lambda.min"))

#Create a tibble of coefficients ordered by absolute value
sig_glms <- tibble(
      codons_glmnet = rownames(coefs),
      coeff_glmnet = coefs[,1]
    ) %>%
        arrange(desc(abs(coeff_glmnet))) %>%
          filter(codons_glmnet != "(Intercept)") %>%
          filter(coeff_glmnet != 0)



########
model_combined <- left_join(model_data_codons, model_data_k_dr, by = "geneName")

  #creates an operator for Not %in%
  '%!in%' <- function(x,y)!('%in%'(x,y))
  
    model_combined_sub <- model_combined %>%
      filter(geneName %!in% list_genes)
    
    model_combined_sel <- model_combined %>%
      filter(geneName %in% list_genes)
    
    pred_matrix_sel <- model_combined_sel %>%
    dplyr::select(-geneName, -log2_hlife)

    

predictors <- paste(c(paste(codons, collapse = " + "),
                  paste(sig_glms_k_dr$motif_glmnet, collapse = " + ")), 
                collapse = " + ")



lm_combined_dr <- lm(data = model_combined, 
                      paste("log2_hlife", predictors, sep = " ~ "))


predictors_sub <- paste(c(paste(codons, collapse = " + "),
                  paste(c("TGTAAATA", "TGCAT", "TTTTTTA", "ATATTC"), collapse = " + ")), 
                collapse = " + ")

step_lm <- lm(data = model_combined_sub, 
                      paste("log2_hlife", predictors, sep = " ~ "))


lm_step_combined <- step(step_lm, scope = list(upper = step_lm, lower = lm_codons_dr), trace = FALSE)


predicted_step_sel <- model_combined_sel %>%
  mutate(pred_hlife = predict(lm_step_combined, 
                              pred_matrix_sel))


predicted_step_sel %>%
  ggplot(aes(x = log2_hlife, y = pred_hlife, col = geneName)) +
    geom_point() + 
    scale_x_continuous(limits = c(0, 5)) + 
    scale_y_continuous(limits = c(0, 5)) + 
    geom_abline() +
#    theme_coding() + 
    labs(y = "Predicted Half life", 
         x = "Measured Half Life")
```

