---
title: "Comparing Predicted and Measured hlife of Candidate Genes"
author: "Abhishek Jain"
date: "18 June 2019"
output: 
  md_document:
    variant: markdown_github
editor_options:
  chunk_output_type: console
---

## Data exploration

```{r setup}

if(!require(tidyverse)){
    install.packages("tidyverse")
    library(tidyverse)
}
if(!require(glmnet)){
    install.packages("glmnet")
    library(glmnet)
}
if(!require(ggplot2)){
    install.packages("ggplot2")
    library(tidyverse)
}

```

```{R Load Data,include=FALSE}

UTR_raw <- read_rds("data/Sun_mutation_UTRs.rds")

#Load Manually created motifs list into a vector
motifs_raw <- scan("data/list_motifs.txt", character())
list_genes <- scan("data/genes_terminators_list.txt", character())

motifs_cheng = c("TGTAAATA", "TGCAT", "TTTTTTA", "ATATTC")

#this is new karsten/Chan et al 2018 data
  dr_raw <- read_tsv("data/new_karsten_dr_data.txt")

  
```

```{R clean data}
#create a vector of 3'UTRs for searching motifs later 
UTR_3 <- UTR_raw$UTR3_seq

#change colnames, add means and filter missing values 
dr_raw2 <- dr_raw %>%
  rename_all(~c("geneName","comgeneName","hlife_r1","hlife_r2")) %>%
    #dplyr::select(-gene) %>%
      mutate(hlife = rowMeans(cbind(hlife_r1, hlife_r2), na.rm = TRUE)) %>% 
        filter(is.finite(hlife)) 

#cleaner df
dr_data <- dr_raw2 %>%
          dplyr::select(geneName, comgeneName, hlife)

dr_data_sub <- dr_data %>%
  filter(geneName %in% list_genes)
          
  
```



Calculate 3'UTR Motif Frequencies from 3'UTR Ref sequences
```{R motif search}

#Dictionary for non-specific codes and converting U -> T
motifs <- motifs_raw %>% str_replace_all(c("U" = "T", "W" = "(A|T)", "S" = "(C|G)", "M" = "(A|C)", "K" = "(G|T)", "R" = "(A|G)", "Y" = "(C|T)", "B" = "(C|G|T)", "D" = "(A|G|T)", "H" = "(A|C|T)", "V" = "(A|C|G)", "N" = "(A|C|G|T)"))

#Initate ref tibble and store gene names
ref_motifs <- tibble(geneName = UTR_raw$genename)


#Search and add frequency of each c(motif) as a column in ref dataset
for (i in 1:length(motifs)){
ref_motifs <- mutate(.data = ref_motifs, 
                      !!motifs_raw[i] := str_count(UTR_3, motifs[i]))
}


```

Add motif frequencies to each dataset

```{R Add Motif Information}

#add dr data for Chan et al
ref_motifs_dr <- left_join(ref_motifs, dr_data, by = "geneName") %>%
  filter(is.finite(hlife))

ref_motifs_dr_sub <- ref_motifs_dr %>%
  filter(geneName %in% list_genes)


#get best motifs out
ref_motifs_dr_sub %>% 
  dplyr::select(-hlife) %>% 
    gather(key = "motif", value = "frequency", -geneName) %>% 
      group_by(motif) %>% 
        summarise(sum = sum(frequency)) %>% 
      ungroup() %>% 
    arrange(desc(sum)) %>% select()


```

```{R Make Data binary}
##Decay Data Karsten/Chan
model_data_k_dr <- ref_motifs_dr %>%
  #convert motif frequencies to presence and then to factor
    mutate_at(motifs_raw, funs(replace(., .>= 1, 1)))%>%
    mutate_at(motifs_raw, funs(factor(.) ))%>%
  #also convert to hlife to log2fold
    mutate(log2_hlife = log2(hlife))%>%
    select(-hlife)

#selected genes from model_data
model_data_sel <- model_data_k_dr %>%
  filter(geneName %in% list_genes)

#Create a subset of model_data for linear modelling

  #creates an operator for Not %in%
  '%!in%' <- function(x,y)!('%in%'(x,y))
  
    model_data_sub <- model_data_k_dr %>%
    filter(geneName %!in% list_genes)
```

```{R Glmnet Matrix}
#Chan et al
motifs_matrix_k_dr <- model_data_k_dr %>%
    dplyr::select(-geneName, -log2_hlife) %>%
      sapply(., strtoi)%>%
        data.matrix()

selected_data <- model_data_k_dr %>%
  filter(geneName %in% list_genes) %>%  
    dplyr::select(-geneName, -log2_hlife)
```

```{Performing Glmnet on Chan et al 2018}

#Cross Validated Glmnet to get lambda.min for extracting coeffs 
cv_glm_k_dr <- cv.glmnet(x = motifs_matrix_k_dr, 
                         y = model_data_k_dr$log2_hlife,
                         nfolds = 100)

#visualize MSE vs log(lambda plot) for CV glmnet
plot(cv_glm_k_dr, sub = "DATA: Chan et al. 2018 ")


```

```{R glmnet Linear Modelling}
coefs_k_dr <- data.matrix(coef(cv_glm_k_dr, s = "lambda.min"))

#Create a tibble of coefficients ordered by absolute value
sig_glms_k_dr <- tibble(
      motif_glmnet = rownames(coefs_k_dr),
      coeff_glmnet = coefs_k_dr[,1]
    ) %>%
        arrange(desc(abs(coeff_glmnet))) %>%
          filter(motif_glmnet != "(Intercept)") %>%
          filter(coeff_glmnet != 0)

#create multiple lm model for predtion 
lm_motifs_k_dr <- lm(data = model_data_sub, 
                     paste("log2_hlife", 
                           paste(sig_glms_k_dr$motif_glmnet, 
                                 collapse = " + "),
                           sep = " ~ "))
```

```{R Prediction}
predict(lm_motifs_k_dr, selected_data)
predicted_data <- model_data_sel %>%
  mutate(pred_log2hlife = predict(lm_motifs_k_dr, selected_data))

predicted_data %>%
  ggplot(aes(x = log2_hlife, y = pred_log2hlife, col = geneName)) +
    geom_point() + 
    scale_x_continuous(limits = c(0, 5)) + 
    scale_y_continuous(limits = c(0, 4)) + 
    geom_abline() +
#    theme_coding() + 
    labs(y = "Predicted Half life", 
         x = "Measured Half Life")

```