---
title: "Abhi Create Reference Data"
author: "Abhishek Jain"
date: "03/09/2019"
output: 
  md_document:
    variant: markdown_github
editor_options:
  chunk_output_type: console
---

#Updates: 
1) Converted motif calculation and codon usage to functions
2) Codon usage is now calculated denovo from CDS_seq instead of taken from rawdata
    (Note: This is slow - 30 sec bottleneck)
3) Codons are not loaded from a list but generated in-code using combination 


## Initation:
```{R Setup}

if(!require(tidyverse)){
    install.packages("tidyverse")
    library(tidyverse)
}
if(!require(ggplot2)){
    install.packages("ggplot2")
    library(tidyverse)
}
if(!require(readxl)){
    install.packages("readxl")
    library(readxl)
}
#theme for ggplot  
  theme_coding <- function(...){
  theme_bw()+
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title = element_text(size = 12),
        panel.grid = element_blank(),
        plot.margin = unit(c(0.5, 0.5, 0.5, 0.5), units = , "cm"),
        plot.title = element_text(size = 20, vjust = 1, hjust = 0.5),
        legend.text = element_text(size = 12),
        legend.title = element_blank(),
        legend.position = c(0.9, 0.9))
}

#defining fun color for scheming
colour <- c(teal = "#237E8A", green = "#1A8A21", red = "#A62621")

```

Important: DR Data used here is karsten weiss only.

```{R Load Data}

#get reference UTRs and CDS sequences from Cheng et al Suppl.
ref_raw <- read_rds("data/Sun_mutation_UTRs.rds")
  #Get coding sequences from ref_raw in a separate vector
  cds_seq <- ref_raw$CDS_seq
  
#Load a motifs list into a vector
motifs_raw <- scan("data/list_motifs.txt", character())

##Loading of codons_raw is commented out because it is now generated in-code
  #codons_raw <- scan("data/codons.txt", character())

#create combinations of codons
codons_raw <- as_tibble(expand.grid(c("A","T","G","C"), 
                                    c("A","T","G","C"), 
                                    c("A","T","G","C"))) %>%
  #combine 3 columns into a single one
  unite(Var, Var1, Var2, Var3, sep = "") %>% 
    #remove three stop codons and convert to a vector
    filter(!Var %in% c("TAG", "TAA", "TGA")) %>% .$Var 

```

```{R Calculate Codon Usage}

#create function to extract and count CU proportional to length of CDS
calc_CU <- function(data, codons, geneID = "genename", CDS = "CDS_seq"){
  
  ref_codons <- data %>% select(geneName = geneID, CDS_seq = CDS)
  
  as_tibble(bind_cols(ref_codons, set_names(lapply(codons, function(each_codon){
    seq_list <- str_extract_all(ref_codons$CDS_seq, "...")
    sapply(seq_list, function(each_vec){
      sum(str_count(each_vec, pattern = each_codon))/length(each_vec)
      })
  }), codons))) %>% select(-TTT, -CDS_seq)
}


    
ref_codons <- calc_CU(ref_raw, codons_raw, geneID = "genename")


```


```{R CU Test}
#sunetal CU
sunetal_codon <- ref_raw %>% as_tibble %>%
  select(genename, CDS_seq, TTT:GGG) %>%
    rename(geneName = genename) %>%
  gather(key = "codon", value = "freq", TTT:GGG) %>%
    mutate(freq = freq/nchar(CDS_seq)) %>%
  spread(codon, freq) %>%
    select(geneName, codons_1) 

sam_codon <- tibble(geneName = ref_raw$genename, ORF = as.character(ref_raw$CDS_seq), length = nchar(ref_raw$CDS_seq)) %>% 
  mutate(ORF = gsub("([ATCG]{3})([ATCG]{3})", '\\1,\\2,', as.character(ORF))) %>% 
  filter((length %% 3) == 0) %>% 
  separate_rows(ORF, sep = ",") %>% 
  group_by(geneName, ORF) %>% 
  summarise(counts = n()) %>%
  spread(key = ORF, value = counts, fill = 0) %>% 
  select(-V1) %>% 
  inner_join(tibble(geneName = ref_raw$genename, geneLength = nchar(ref_raw$CDS_seq)), by = "geneName") %>%
  gather(key = codon, value = number, -geneLength, -geneName) %>%
  mutate(number = number/geneLength) %>%
  spread(key = codon, value = number) %>%
  select(-TTT)
            



best_codon <- ref_raw %>% 
  as_tibble() %>% 
  select(genename, CDS_seq) %>% 
  mutate(split_CDS = gsub("([ATCG]{3})([ATCG]{3})", '\\1,\\2,', 
                          as.character(CDS_seq))) %>%
  bind_cols(., setNames(lapply(codons_raw, 
                               function(each_codon){ 
                                 stringr::str_count(.$split_CDS, 
                                                    each_codon)/nchar(.$CDS_seq)}),
                        codons_raw)
            ) %>% 
  select(-CDS_seq, -split_CDS, -TTT)

```

```{R Calculate Motif Frequencies}
#count_motifs function takes a reference dataframe as object and a list of motifs 
#it also optionally accepts the names of geneName and 3'UTR sequence columns as string 

count_motifs <- function(genes_UTR, names_motifs, 
                         geneID = "genename", UTR_ID = "UTR3_seq"){
  
  #Dictionary for single letter codes 
    motifs_regex <- names_motifs %>% 
      str_replace_all(c("U" = "T", "W" = "(A|T)", "S" = "(C|G)", "M" = "(A|C)", 
                        "K" = "(G|T)", "R" = "(A|G)", "Y" = "(C|T)", 
                        "B" = "(C|G|T)", "D" = "(A|G|T)", "H" = "(A|C|T)", 
                        "V" = "(A|C|G)", "N" = "(A|C|G|T)"
                        )
                      )
  #clean df_UTR
  genes_UTR <- as_tibble(select(genes_UTR, geneName = geneID, UTR3_seq = UTR_ID)) 
  
  bind_cols(genes_UTR, setNames(lapply(motifs_regex, 
                                     function(each_motif){
                                       stringr::str_count(genes_UTR$UTR3_seq,
                                                          each_motif)}), 
                                names_motifs)) %>% 
    select(-UTR3_seq)

  } #outputs a df with counts for each motif

#Run count motifs function on ref_raw dataset
ref_motifs_all <- count_motifs(ref_raw, motifs_raw)

#filter out a list of motifs with >5 occurances in the 3'UTRome 
motifs_fil <- motifs_raw[sapply(ref_motifs_all[-1], sum) >= 5]

#filter reference motifs data to motifs with >5 occurances
ref_motifs_fil <-  ref_motifs_all %>% 
  select_at(vars(geneName, motifs_fil))
  

```

Combine reference Dataframes 

```{R Combine Codons and Motifs into a single ref_data}

#model df containing motifs and log2(hlife)
ref_data <- left_join(ref_motifs_fil, ref_codons, by = "geneName")

```

Export Data into a CSV

```{R Export Reference Data}
write_csv(ref_data, path = "data/ref_data.csv")

```