---
title: "project summary"
author: "Hanqin Du"
date: "2020/2/1"
output: html_document
---


```{r set work space for r, eval=FALSE, echo=FALSE}
# set work space for Rstudio
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))
```

```{r echo=FALSE}
# KNN validation and cross Validation are extremely time-consuming processes
# set the flags to FALSE to apply previous optimal value and skip the process
eval_KNN_validation = FALSE
eval_crossValidation = FALSE
```

# project discription
Gene expression is the process through which cells generate diversity and resilience to environmental changes. It consists of two steps: transcription, where an intermediate molecule (mRNA) is produced from the DNA, and translation of mRNA into protein. The rate of transcription and translation can be regulated by regulatory elements acting in both cis and trans. 3'UTR, 5'UTR, and promoter are known as powerful regulatory processes that determine the rate of mRNA synthesis and decay.

In this project, we are interested in using these cis-regulatory features to predict multidimensional output that indicates how the mRNA abundance change in multiple environmental conditions. Since our goal is to quantify the effect of different elements, we should avoid clustering and go straight from element to expression pattern. Linear models with multidimensional lasso (known as an algorithm of multi-task learning) which is already implemented in the glmnet package should be a good start. As an extension, we may apply to multiple states of mRNA processing like splicing, decay, and translation by combining multiple data sets. Finally, we can carry out a quantitative analysis of the contribution of motifs in different types of CRE and ask which regions make more of a contribution to dynamic changes in gene expression.



# Index
- ### Part 1: setup and investigate the dataset

  - [Load library and function](#loadFunction)
  - [Construct the design matrix](#constructDesignMatrix)
  - [1868 gene expression level data loss after inner join above data by gene name](#geneLoss)
  - [apply fix-length downstream sequence from fungiDB as 3'UTR for 1320 of the 1868 genes](#applyFungiDB)
  - [apply fix-length downstream sequence from Samuel as 3'UTR for 503 of the 548 genes](#applySamuel)
  - [check the motif frequency from both dataset "cheng" and "fungiDB"](#fungiDBMotifFrequency)
  - [compare motif frequency of 120 downstream and actual 3'UTR](#120downstreamUTRcomparison)
  - [8 motifs are removed from list considering their low frequency on 3’UTR](#removeLowFrequencyMotifs)
  

- ### Part 2: Assess the rationality of group lasso

  - [4 significant motifs on heat-shock-relevant environmental stress](#significantMotifsHS)
  - [9 heat-shock-relevant motifs are selected from linear models](#9significantMotifsHS)
  - [Identify potential relationships between tasks](#potentialRelationshipTask)

- ### Part 3: Justify group Lasso

  - dealing with missing value
    - [distribution of missing value](#distributionMissingValue)
    - [filter out gene with high value-missing ratio](#filterGene)
    - [fill missing value by kNN imputation](#imputation)
    - [baseline: replace missing value with column's mean](#imputationBaseline)
  - fitting models
    - [choose lambda from 10-fold cross validation](#crossValidation)
    - [training model with selected lambda](#trainingModels)
  - evaluate models
    - [regulatory effect of motifs decrease as time pass in heat shock](#regulatoryEffectDecrease)
    
- ### Part 4: analyze the regression coefficients

  - comparison between L2 norm and sum of the non-zero regression coefficients
    - [divide environmental condition into groups](#environmentalGroup)
    - [under all condition](#L2Allcondition)
    - [under heat shock and hypothermia](#heatShockHypothermia)
    - [under heat shock with and without sorbitol](#heatShocksorbitol)
    - [investigate the transcript frequency of each motif against their coefficient](#frequencyCoefficient)
  - [box graph for cross comparison between two type of environmental stress](#boxPlot)
  - [comparison motifs rank between each group of models](#motifRank)
  - [investigating bias and expression mean of each model](#bias)
  - [Select motif with weighted square L2](#weightedL2)
  - [Compare mean of L2 of significant motifs across different group of environmental stress](#meanL2)
  - [compare proportion of variance explain by each motif](#proportionVar)
  - [significant test with linear model](#linearSignificant)
  



# Part 1: setup and investigate the dataset
After loading all the required library and function, we investigate and evaluate the dataset we have.

### <a id="loadFunction"> Load library and function </a>

```{r load library, results='hide', message=FALSE}
library(data.table)
library(tidyverse)
library(lmodel2)
library(splitstackshape)
library(gridExtra)
library(glmnet)
library(impute)
```


```{r load function, echo=FALSE}

# obtain mean square error from a linear model
# @model: linear model from stats v3.6.2
mse <- function(model){
  mean(summary(model)$residuals^2)
}

# plot graph with error bar by value and deviation
# @dataset{dataframe}: dataset with label, mean and deviation
plot_mean_deviation <- function(dataset,label,mean,deviation){
  plot(1:dim(dataset)[1],dataset[[mean]], pch=19,xlab="",ylab="coefficient",xaxt="n",xlim = c(0.5,dim(dataset)[1]+0.5),
       ylim=c(min(dataset[mean]-1.96*dataset[deviation]),max((dataset[mean]+1.96*dataset[deviation]))))
  lines(rbind(1:dim(dataset)[1],1:dim(dataset)[1],NA),rbind(dataset[[mean]]-1.96*dataset[[deviation]],dataset[[mean]]+1.96*dataset[[deviation]],NA))
  axis(side=1,at=1:dim(dataset)[1],labels=dataset[[label]])
}

# extract coefficients from selected mgaussian model
# @models: mguassian model from glmnet
# @conditions: list of names of sub model (e.g. hs_25Cto33C_10min)
extract_coefficient <- function(models, conditions) {
  coefficients = NULL
  
  for (condition in conditions) {
    column = numeric(models$beta[[condition]]@Dim[1])
    column[models$beta[[condition]]@i+1] = models$beta[[condition]]@x
    coefficients = cbind(coefficients,column)
  }
  coefficients = as.data.frame(coefficients)
  names(coefficients) = conditions
  row.names(coefficients) = names_motifs_valid
  return(melt(setDT(coefficients, keep.rownames = TRUE), "rn"))
}

# plot selected models coefficient with line graph
# @models: mguassian model from glmnet
# @conditions: list of names of sub model (e.g. hs_25Cto33C_10min)
# @lineNumber: indicate how many factors should be includes in each plot
plot_models_line <- function(models, conditions, lineNumber) {
  coefficients = NULL
  
  for (condition in conditions) {
    column = numeric(models$beta[[condition]]@Dim[1])
    column[models$beta[[condition]]@i+1] = models$beta[[condition]]@x
    coefficients = cbind(coefficients,column)
  }
  coefficients = as.data.frame(coefficients)
  names(coefficients) = conditions
  row.names(coefficients) = names_motifs_valid
  melt(setDT(coefficients, keep.rownames = TRUE), "rn")
  
  i = 1
  while (i <= dim(coefficients)[1]) {
    if((i+lineNumber)<= dim(coefficients)[1]){
      print(
      ggplot(data = melt(setDT(coefficients[i:(i+lineNumber-1)], keep.rownames = TRUE), "rn"), aes(x=variable,y=value,group = rn)) + 
        geom_line(aes(colour=rn), size = 1.7)+
        labs(x = "model", y = "regression coefficient") +
        theme(axis.text.x = element_text(size = 16, angle = 60, hjust = 1),axis.text.y = element_text(size = 16),axis.title=element_text(size=16,face="bold"))
      )
    }else{
      print(
      ggplot(data = melt(setDT(coefficients[i:dim(coefficients)[1]], keep.rownames = TRUE), "rn"), aes(x=variable,y=value,group = rn)) + 
        geom_line(aes(colour=rn), size = 1.7)+
        labs(x = "model", y = "regression coefficient") +
        theme(axis.text.x = element_text(size = 16, angle = 60, hjust = 1), axis.text.y = element_text(size = 16),axis.title=element_text(size=16,face="bold"))
      )
    }
    
    i = i + lineNumber
  }
  
}

# extract all coefficients from a linear model, return a dataframe with an additional column with valued @groupName
# @model: linear model from stats v3.6.2
extract_coefficient_lm <- function(model, groupName){
  df = data.frame(summary(model)$coefficients[2:model$rank,])
  df$motif = rownames(df)
  df$condition = groupName
  rownames(df) = NULL
  return(df)
}

```


### <a id="constructDesignMatrix"> Construct the design matrix </a>

##### Expression level change of 6152 distinct genes under 173 different environmental stresses from Gasch's study
In this project, we imported the data describes gene expression level in various environment condition from Gasch's study where DNA microarrays were applied to measure the relative abundance of mRNA (relative transcript level) before and after a series of environmental stresses. The dataset consists of 6152 rows and 176 columns where each row represents a gene of yeast. The first 3 columns provide the details of the gene and the other 173 columns describe the normalized, background-corrected log2 values expression level change under 173 different environmental stress.


##### The 69 candidate sequence motifs we use in this project is from 3 different studies:

53 candidate sequence motifs came from the study of Shalgi et al (2005), where motifs are derived by analyzing the exhaustively enumerating all k-mers(k = {8,9,10,11,12}) and looking for over-represented motifs with extreme half-life value. 515 significant k-mers were selected with ranksum test and clustered by ClustalW which resulted in 51 clusters of motifs. Additionally, two more motifs were found by grouping genes with extreme half-life and ran Gibbs sampler.

14 candidate sequence motifs came from the study of Hogan et al. (2008). Two related computational methods were applied to identify candidate binding sites of 40 out of the more than 500 known and predicted RBPs in S. cerevisiae: (1)“finding informative regulatory elements” (FIRE) and (2)“relative filtering by nucleotide enrichment” (REFINE). The former searches for motifs with informative patterns of enrichment. The other one identifies all hexamers that are significantly enriched in untranslated regions, filters out regions of target sequences that are relatively devoid of such hexamers, and then applies the "multiple expectation maximization for motif elicitation" (MEME) motif-finding algorithm. As a result, 14 motifs that are likely to be the binding sites of 16 RBP are found. 

The rest 4 were from the study of Cheng (2017): four mRNA-stability related motifs in the 3′ UTR were found by De novo motif searching and their reliability and effect have been examined by linear mixed effect model, Fisher test P-value corrected with Benjamini–Hochberg, Wilcoxon ranksum test and multivariate linear regression.


##### UTR sequence of 4388 genes from Cheng et al. 
The majority of 3'UTR data came from the research of Cheng et al. Annotated transcript sequences data of 4284 genes out of 6152 genes were downloaded from the GitHub sites introduced in their paper.

```{r echo=FALSE, warning=FALSE, message=FALSE}

# load Cheng's UTR sequence 
UTR_cheng <- read_rds("../data/Sun_mutation_UTRs.rds")


# load motifs and covert to regular expression
motifs_raw <- scan("../data/list_motifs.txt", character())

motifs <- motifs_raw %>% str_replace_all(c("U" = "T", "W" = "(A|T)", "S" = "(C|G)", "M" = "(A|C)", "K" = "(G|T)", "R" = "(A|G)", "Y" = "(C|T)", "B" = "(C|G|T)", "D" = "(A|G|T)", "H" = "(A|C|T)", "V" = "(A|C|G)", "N" = "(A|C|G|T)"))


# load Gasch's data and convert all the expression data to number
expressionLevel_Gasch <- read_tsv("../data/Gasch2000_complete_dataset_rename.txt", 
                       locale = locale(decimal = ","))

for (i in names(expressionLevel_Gasch)[3:176]) {
  expressionLevel_Gasch[[i]] <- as.numeric(as.character(expressionLevel_Gasch[[i]]))
}

```

```{r echo=FALSE}

# initate ref tibble and store gene names
ref_motifs <- tibble(geneName = UTR_cheng$genename)

# search and add frequency of each c(motif) as a column in ref dataset
for (i in 1:length(motifs)){
  ref_motifs <- mutate(.data = ref_motifs,!!motifs_raw[i] := str_count(UTR_cheng$UTR3_seq, motifs[i]))
}


# record all names of motifs and environmental stress
names_motifs_all = names(ref_motifs)[2:length(ref_motifs)]
names_environmental_stress = names(expressionLevel_Gasch)[4:length(expressionLevel_Gasch)]

# construct design matrix by inner join expression data and motif frequency data
fullTable_Gasch <- merge(expressionLevel_Gasch,ref_motifs,by = "geneName")

```


### <a id="geneLoss"> 1868 gene expression level data loss after inner join above data by gene name. </a>
After inner joining the expression level of 6152 genes with the 4388 motif frequency data which computed from 3'UTR data of Cheng's study, expression level data of 1868 genes loss.
```{r echo=FALSE}
names_gene_information_loss = setdiff(expressionLevel_Gasch$geneName,UTR_cheng$genename)

print(paste("number of missing 3'UTR sequence", length(names_gene_information_loss)))
```


### <a id="applyFungiDB">apply fix-length downstream sequence from fungiDB as 3'UTR for 1320 of the 1868 genes. </a>
First, we checked the mean length of 3'UTR from Cheng et al.
```{r echo=FALSE}
print(paste("median length of 3'UTR from Cheng et al. = ", median(UTR_cheng$UTR3_length,na.rm = TRUE)))
print(paste("mean length of 3'UTR from Cheng et al. = ", mean(UTR_cheng$UTR3_length,na.rm = TRUE)))
```

Since there is no reliable 3’UTR annotation in most of the online database, we applied downstream sequence with fix-length (120 nucleotides) as 3’UTR. The length of 120 nucleotides is considered as a sensible number close to the median length.
```{r warning=FALSE}
# load UTR sequence from fungiDB
UTR_fungiDB = read_csv("../data/3UTR_fungiDB")
UTR_Cheng_fungiDB = rbind(UTR_cheng,UTR_fungiDB,fill=TRUE)

names_gene_cheng = UTR_cheng$genename
names_gene_fungiDB = UTR_fungiDB$genename
```


### Construct the design matrix with additional data from fungiDB
```{r}
# get UTR sequence from both dataset

ref_motifs <- tibble(geneName = UTR_Cheng_fungiDB$genename)
for (i in 1:length(motifs)){
  ref_motifs <- mutate(.data = ref_motifs,!!motifs_raw[i] := str_count(UTR_Cheng_fungiDB$UTR3_seq, motifs[i]))
}


# merge motifs
fullTable_Gasch <- merge(expressionLevel_Gasch,ref_motifs,by = "geneName")

```



### <a id="fungiDBMotifFrequency"> check the motif frequency from both dataset "cheng" and "fungiDB" </a>
```{r}

fullTable_Gasch_original = fullTable_Gasch[fullTable_Gasch$geneName %in% names_gene_cheng,]
fullTable_Gasch_fungiDB = fullTable_Gasch[fullTable_Gasch$geneName %in% names_gene_fungiDB,]


summary_motif_frequency = data.frame(matrix(ncol = 3, nrow = 0))

for(m in names_motifs_all){
  summary_motif_frequency = rbind(summary_motif_frequency, data.frame(
    motif = m,
    dataset = "Cheng",
    frequency = sum(fullTable_Gasch_original[[m]]),
    frequency_per_gene = sum(fullTable_Gasch_original[[m]])/(dim(fullTable_Gasch_original)[1])
  ))
  summary_motif_frequency = rbind(summary_motif_frequency, data.frame(
    motif = m,
    dataset = "fungiDB",
    frequency = sum(fullTable_Gasch_fungiDB[[m]]),
    frequency_per_gene = sum(fullTable_Gasch_fungiDB[[m]])/(dim(fullTable_Gasch_fungiDB)[1])
  ))
}


```


```{r echo=FALSE, fig.height=8, fig.width=14}
summary_motif_frequency = summary_motif_frequency[order(-summary_motif_frequency$frequency_per_gene),]
summary_motif_frequency_temp = summary_motif_frequency[summary_motif_frequency$motif %in% head(unique(summary_motif_frequency$motif),35),]

ggplot(data=summary_motif_frequency_temp, aes(x=reorder(motif,-frequency_per_gene), y=frequency_per_gene, fill=dataset)) +
  geom_bar(stat="identity", position=position_dodge()) +
  theme(axis.text.x = element_text(size = 16, angle = 70, hjust = 1),
        axis.text.y = element_text(size = 16),
        legend.text = element_text(size = 16),
        axis.title = element_text(size = 16)) + 
  scale_x_discrete(name ="Motif") + 
  ylim(0,0.35)+
  ggtitle("frequency of each motif per gene according to the dataset from Cheng's study and fungiDB")
```

```{r echo=FALSE, fig.height=8, fig.width=14}
summary_motif_frequency = summary_motif_frequency[order(-summary_motif_frequency$frequency_per_gene),]
summary_motif_frequency_temp = summary_motif_frequency[summary_motif_frequency$motif %in% tail(unique(summary_motif_frequency$motif),34),]

ggplot(data=summary_motif_frequency_temp, aes(x=reorder(motif,-frequency_per_gene), y=frequency_per_gene, fill=dataset)) +
  geom_bar(stat="identity", position=position_dodge()) +
  theme(axis.text.x = element_text(size = 16, angle = 70, hjust = 1),
        axis.text.y = element_text(size = 16),
        legend.text = element_text(size = 16),
        axis.title = element_text(size = 16)) + 
  scale_x_discrete(name ="Motif") + 
  ylim(0,0.35)+
  ggtitle("frequency of each motif per gene according to the dataset from Cheng's study and fungiDB")
```




### <a id="120downstreamUTRcomparison">compare motif frequency of 120 downstream and actual 3'UTR</a>

```{r echo=FALSE, warning=FALSE, message=FALSE}

#ref datasets for UTRs 
UTR_SUN_fungiDB = read_csv("../data/3UTR_SUN_fungiDB")

#Get sequences from UTR_all in a separate vector
UTR_3 <- UTR_SUN_fungiDB$UTR3_seq

#Load Manually created motifs list into a vector
motifs_raw <- scan("../data/list_motifs.txt", character())
motifs_cheng = c("TGTAAATA", "TGCAT", "TTTTTTA", "ATATTC")


```


##### Construct Motif Frequencies Matrix from 3'UTR Ref sequences
```{r echo=FALSE}

#Dictionary for non-specific codes and converting U -> T
motifs <- motifs_raw %>% str_replace_all(c("U" = "T", "W" = "(A|T)", "S" = "(C|G)", "M" = "(A|C)", "K" = "(G|T)", "R" = "(A|G)", "Y" = "(C|T)", "B" = "(C|G|T)", "D" = "(A|G|T)", "H" = "(A|C|T)", "V" = "(A|C|G)", "N" = "(A|C|G|T)"))

#Initate ref tibble and store gene names
ref_motifs_120downstream <- tibble(geneName = UTR_SUN_fungiDB$genename)


#Search and add frequency of each c(motif) as a column in ref dataset
for (i in 1:length(motifs)){
  ref_motifs_120downstream <- mutate(.data = ref_motifs_120downstream,!!motifs_raw[i] := str_count(UTR_3, motifs[i]))
}

names_motifs_all = names(ref_motifs)[2:length(ref_motifs)]
```


```{r}
ref_motifs = ref_motifs[ref_motifs$geneName %in% ref_motifs_120downstream$geneName,]
```


##### plot the frequency per gene
```{r echo=FALSE}
summary_motif_frequency = data.frame(matrix(ncol = 3, nrow = 0))

for(m in names_motifs_all){
  summary_motif_frequency = rbind(summary_motif_frequency, data.frame(
    motif = m,
    dataset = "3_UTR",
    frequency_per_gene = sum(ref_motifs[[m]])/nrow(ref_motifs)
  ))
  summary_motif_frequency = rbind(summary_motif_frequency, data.frame(
    motif = m,
    dataset = "120_downstream",
    frequency_per_gene = sum(ref_motifs_120downstream[[m]])/nrow(ref_motifs_120downstream)
  ))
}

```

```{r echo=FALSE, fig.height=8, fig.width=14}
summary_motif_frequency = summary_motif_frequency[order(-summary_motif_frequency$frequency_per_gene),]
summary_motif_frequency_temp = summary_motif_frequency[summary_motif_frequency$motif %in% head(unique(summary_motif_frequency$motif),35),]

ggplot(data=summary_motif_frequency_temp, aes(x=reorder(motif,-frequency_per_gene), y=frequency_per_gene, fill=dataset)) +
  geom_bar(stat="identity", position=position_dodge()) +
  theme(axis.text.x = element_text(size = 16, angle = 70, hjust = 1),
        axis.text.y = element_text(size = 16),
        legend.text = element_text(size = 16),
        axis.title = element_text(size = 16)) + 
  scale_x_discrete(name ="Motif") + 
  ylim(0,0.35)+
  ggtitle("comparison of the frequency of each motif per gene between actual 3'UTR and 120 downstream")
```




### <a id="applySamuel"> Construct the design matrix with additional UTR sequence data from Samuel </a>
additional 503 downstream sequences from Samuel's lab dataset are added to construct the design matrix

```{r}
UTR_Samuel <- read.csv(file="../data/genome_wide_120nt_3UTR_sequence.csv")
colnames(UTR_Samuel) = c("genename","UTR3_seq")

UTR_Cheng_fungiDB_Samuel = rbind(UTR_Cheng_fungiDB,UTR_Samuel,fill=TRUE)
```

```{r}
# get UTR sequence from both dataset
UTR_3 <- UTR_Cheng_fungiDB_Samuel$UTR3_seq

# Search and add frequency of each c(motif) as a column in ref dataset
ref_motifs <- tibble(geneName = UTR_Cheng_fungiDB_Samuel$genename)
for (i in 1:length(motifs)){
  ref_motifs <- mutate(.data = ref_motifs,!!motifs_raw[i] := str_count(UTR_3, motifs[i]))
}


# merge motifs
fullTable_Gasch <- merge(expressionLevel_Gasch,ref_motifs,by = "geneName")

```



### <a id="removeLowFrequencyMotifs"> 8 motifs are removed from list considering their low frequency on 3'UTR </a>
We matching the 69 motifs on every gene sequence to compute the frequency matrix which could be used as the design matrix of our models. The distribution of the frequency of each motif across all the genes is very unbalanced. 8 motifs have never been investigated from any of the 3'UTR from the study of Cheng et al. Therefore, we temporarily remove them.

| Frequency  | number of motifs (n) |
| ---------- | -------------------- |
| n = 0      | 7                    |
| 0 < n < 6  | 8                    |
| 5 < n < 21 | 7                    |
| n > 20     | 47                   |

One possible explanation of the huge amount of low-frequency motifs is that it is hard to match exactly the regular expression to the UTR sequence. Another point is, the secondary-structure motifs are more likely to be found on 5'UTR rather than 3'UTR. Also, a motif with a significant effect on regulation is reasonable to have a small frequency. 


```{r echo=FALSE}
# construct the motif frequency table
motifs_count_sum <- colSums(fullTable_Gasch[names_motifs_all],na.rm=TRUE)

# plot histgram
hist(motifs_count_sum,  main = 'histogram of motifs frequency sum', breaks = 50)

# remove zero-frequency motifs from list
remove_list = NULL
for (i in names(motifs_count_sum)){
  if (motifs_count_sum[[i]] == 0){
    remove_list <- c(remove_list,i)
  }
}
names_motifs_valid <- names_motifs_all[!names_motifs_all %in% remove_list]
```

```{r include=FALSE}
length(motifs_count_sum[motifs_count_sum==0])
length(motifs_count_sum[motifs_count_sum>0 & motifs_count_sum<6])
length(motifs_count_sum[motifs_count_sum>5 & motifs_count_sum<21])
length(motifs_count_sum[motifs_count_sum>20 & motifs_count_sum<51])
length(motifs_count_sum[motifs_count_sum>50])
```







# Part 2: Assess the rationality of group lasso
Before applying multi-task learning, we would like to investigate the data and assess if it is reasonable to apply group lasso on them.


### <a id="significantMotifsHS"> 4 significant motifs on heat-shock-relevant environmental stress.</a>
To investigate how the expression level changes linearly depends on the motifs, we calculate the Pearson correlation coefficient (PCC) between them. Pearson correlation coefficient is a measure of the linear correlation between two variables X and Y:
$$
p_{x,y}= {cov(X,Y)\over \sigma_X \sigma_Y}
$$
We start by picking a group of environmental conditions about Heat Shock from Various Temperatures to 37°C. Four motifs show a much higher correlation coefficient rather than the others, which, indicate the linear relationship between these motifs and expression level. One point worth mentioning is that all these motifs are also focused by Abhi where `TGTATAWT` is explained to be positively associated with RNA stability and the rest three of them are explained to be negatively associated with the RNA stability. However, Abhi is not quite confident with the conclusion of `UGUAHMNUA` as the negative relationship can only be derived from Chan's data but not Sun's data. 

```{r}
names_temperature_condition = c(
  "hs_17to37_20min",
  "hs_21to37_20min",
  "hs_25to37_20min",
  "hs_29to37_20min",
  "hs_33to37_20min"
)
```

```{r echo=FALSE}
# compute correlation coefficient
motif_overview <- as.data.frame(
  cor(fullTable_Gasch[names_motifs_valid],fullTable_Gasch[names_temperature_condition], use = "na.or.complete"))

# compute mean
motif_overview$R_mean <- rowMeans(motif_overview[names_temperature_condition[1:5]])

```

```{r, echo=FALSE, fig.width=10, fig.height=7}
# plot motifs with high correlation coefficient
ggplot(data = melt(setDT(as.data.frame(motif_overview[c('UGUAHMNUA','TGTATAWT','ATATTC','TGTAAATA'),names_temperature_condition]), keep.rownames = TRUE), "rn"), aes(x=variable,y=value,group = rn)) + 
  geom_line(aes(colour=rn),size=2) + 
  ylim(-0.25,0.1) + 
  labs(x = "heat shock to 37C from", y = "correlation coefficient") +
  scale_x_discrete(labels=c("hs_17to37_20min" = "17C", "hs_21to37_20min" = "21C","hs_25to37_20min" = "25C","hs_29to37_20min" = "29C", "hs_33to37_20min" = "33C")) + 
  theme(text = element_text(size=24))
```



### <a id="9significantMotifsHS"> 9 heat-shock-relevant motifs are selected from linear models </a>

Before fitting the linear model, we temporarily remove the motifs with a total frequency lower than 5 since we are not confident enough to tell whether a low-frequency motif depends linearly on the heat-shock related environment condition. Then, we fitted the linear model with the `lm()` function with default least squares method:
$$
\min_w \sum (w*e)^2
$$

```{r echo=FALSE}
# filter out motif with frequency lower than or equal to 5
remove_list = NULL

for (i in names(motifs_count_sum)){
  if (motifs_count_sum[[i]] < 6){
    remove_list <- c(remove_list,i)
  }
}

names_motifs_valid <- names_motifs_all[!names_motifs_all %in% remove_list]
```


```{r echo=FALSE}
# linear regression on heat shock from 17C to 37C
formula_model_hs_17to37 <- as.formula(paste("hs_17to37_20min ~",paste(names_motifs_valid, collapse = "+")))
model_hs_17to37 = lm(formula_model_hs_17to37,fullTable_Gasch)

par(mfrow=c(2,2))
plot(model_hs_17to37)
print(paste("mean square error:",mse(model_hs_17to37)))

```


 9 heat-shock-relevant motifs are selected from the linear model with the following coefficient:
 
 
| Motif      | Estimate | Std.Error | t value | Pr(>\|t\|) |
| ---------- | -------- | --------- | ------- | ---------- |
| UGUAHMNUA  | -1.054   | 0.085     | -12.369 | <2e-16     |
| ATATTC     | 0.309    | 0.057     | 5.432   | 1.54e-14   |
| TGTATAWT   | -0.468   | 0.095     | -4.907  | 5.07e-10   |
| WUUGUAWUWU | -0.534   | 0.157     | -3.413  | 0.000207   |
| UAAUAAUW   | -0.250   | 0.083     | -3.025  | 0.106042   |
| TGTAAATA   | 0.244    | 0.106     | 2.308   | 0.000349   |
| AAAATAAAG  | -0.568   | 0.281     | -2.018  | 0.022646   |
| UUUAAUGA   | 0.302    | 0.162     | 1.864   | 0.000394   |
| TCATGTAT   | -0.327   | 0.207     | -1.580  | 0.1141     |


This table is sorted by  `Pr(>|t|)` (also known as the p-value) which is the probability of achieving a |t| as large as or larger than the observed absolute t value if the null hypothesis (estimate = 0) was true. In short, it is the probability that there is no relationship between that feature and the predicted value (Ronald L. Wasserstein et al., 2016). `Estimate` shows the coefficient or weight gains by the responding features in this linear model. `Std.Error` is the standard error of the `estimate` value, which, can be used to calculate the Confidence interval. For example, the 95% confidence interval could be obtained from Estimate +- 1.96*Std.Error. `t value` is calculated from the estimates divided by their standard errors. To make the best use of this value, we need to look up the table of t distribution to learn the reject boundary. In this case, we could simply say the larger the magnitude of the t-value is, the less likely that the coefficient is 0.



### <a id="potentialRelationshipTask"> Identify potential relationships between tasks </a>

To penalty the irrelevant features and ensure the comparison across the models, we plan to apply group lasso. However, before applying the multi-task learning method, we want to ensure there are potential relationships between tasks.

```{r}
names_motifs_temperature_significant = c('UGUAHMNUA','TGTATAWT','ATATTC','TGTAAATA')

names_temperature_condition = c(
  "hs_15min_hs-1",
  "hs_30min_hs-1",
  "hs_40min_hs-1",
  "hs_60min_hs-1",
  "hs_80min_hs-1"
)
```

```{r, echo=FALSE, fig.width=10, fig.height=7}

# compute correlation coefficient
motif_overview <- as.data.frame(
  cor(fullTable_Gasch[names_motifs_valid],fullTable_Gasch[names_temperature_condition], use = "na.or.complete"))

# plot motifs with high correlation coefficient
ggplot(data = melt(setDT(as.data.frame(motif_overview[names_motifs_temperature_significant,names_temperature_condition]), keep.rownames = TRUE), "rn"),
       aes(x=variable,y=value,group = rn)) + 
  geom_line(aes(colour=rn),size = 2) + 
  ylim(-0.3,0.1) + 
  labs(x = "after heat shock", y = "correlation coefficient") +
  scale_x_discrete(labels=c("hs_15min_hs-1" = "15min", "hs_30min_hs-1" = "30min","hs_40min_hs-1" = "40min","hs_60min_hs-1" = "60min","hs_80min_hs-1" = "80min")) + 
  theme(text = element_text(size=24))
```

```{r}
# pick a group of temperature condition
names_temperature_condition = c(
  "29C(1M_sorbitol)~33C(1M_sorbitol)_05min",
  "29C(1M_sorbitol)~33C(1M_sorbitol)_15min",
  "29C(1M_sorbitol)~33C(1M_sorbitol)_30min"
)
```

```{r echo=FALSE, fig.width=10, fig.height=7}
# compute correlation coefficient
motif_overview_steady_temperature <- as.data.frame(
  cor(fullTable_Gasch[names_motifs_valid],fullTable_Gasch[names_temperature_condition], use = "na.or.complete"))

# plot
ggplot(data = melt(setDT(as.data.frame(motif_overview_steady_temperature[names_motifs_temperature_significant,names_temperature_condition]), keep.rownames = TRUE), "rn"),
       aes(x=variable,y=value,group = rn)) + geom_line(aes(colour=rn),size = 2) + ylim(-0.25,0.1) + 
  labs(x = "after heat shock with sorbitol", y = "correlation coefficient")+
  scale_x_discrete(labels=c("29C(1M_sorbitol)~33C(1M_sorbitol)_05min" = "5min", "29C(1M_sorbitol)~33C(1M_sorbitol)_15min" = "15min","29C(1M_sorbitol)~33C(1M_sorbitol)_30min" = "30min")) + 
  theme(text = element_text(size=24))
```


```{r}
names_temperature_condition = c(
  "hs_37to25_15min",
  "hs_37to25_30min",
  "hs_37to25_45min",
  "hs_37to25_60min",
  "hs_37to25_90min"
)
```


```{r, echo=FALSE, fig.width=10, fig.height=7}
# compute correlation coefficient
motif_overview <- as.data.frame(
  cor(fullTable_Gasch[names_motifs_valid],fullTable_Gasch[names_temperature_condition], use = "na.or.complete"))

## plot motifs with high correlation coefficient
ggplot(data = melt(setDT(as.data.frame(motif_overview[names_motifs_temperature_significant,names_temperature_condition]), keep.rownames = TRUE), "rn"),
       aes(x=variable,y=value,group = rn)) + geom_line(aes(colour=rn),size=2) + ylim(-0.1,0.2) + 
  labs(x = "after hypothermia from 37C to 25C", y = "correlation coefficient")+
  scale_x_discrete(labels=c("hs_37to25_15min" = "15min", "hs_37to25_30min" = "30min","hs_37to25_45min" = "45min","hs_37to25_60min" = "60min","hs_37to25_90min" = "90min")) + 
  theme(text = element_text(size=24))

```



# Part 3: apply group lasso

### introduction

To investigate and quantify how the cis-regulatory element affects the gene regulation, we built a series of linear models where each model takes the frequencies of cis-regulatory elements as input and predicts how the gene expression level under certain environmental stress. Their coefficient can be used to estimate the regulatory effect of both a single cis-regulatory element or all the cis-regulatory as a group. For example, if an element is given a relatively large coefficient in a model which predicts the change of expression level under certain environmental stress, we can say the element play a relatively important role in the regulation under that environmental stress. On the other hand, the bias of the model can be considered as the expected gene expression level change of a gene with no cis-regulatory element we considered in the model. In other words, the bias indicates the expression level change that cannot be explained by the model and can be used to estimate the proportion of the efforts on regulation these cis-regulatory elements make. 

At this stage, our aim is to apply group lasso as a multi-task learning method to training all the models together which allows the comparison across each model and ensures their generalization. The L2 penalty term can filter out most of the irrelevant factors by penalizing their coefficient towards zero. 

The penalty on the coefficient vector for variable j is:
$$
(1-\alpha)/2||\beta_j||_2^2+\lambda||\beta_j||_2.
$$


### <a id="distributionMissingValue"> distribution of missing value </a>
Above all the expression level data, around 3% of them are missing value, Where most of the genes (94.8%) have a relatively complete expression levels' data (90%) while 41 genes (0.7%) have missing valves under more than quarter of environmental stresses.


| number of genes | missing value (n) |
| --------------- | ----------------- |
| 755 (12.3%)     | 0                 |
| 4372 (71.1%)    | 0 < n < 9 (5%)    |
| 701 (11.4%)     | 8 < n < 18 (10%)  |
| 283 (4.6%)      | 17 < n < 44 (25%) |
| 41 (0.7%)       | 43 < n < 99 (58%) |


```{r include=FALSE}
sum(is.na(expressionLevel_Gasch[4:176])) #number of missing values
sum(is.na(expressionLevel_Gasch[4:176]))/(173*6152) #ratio of missing value

na_count_sum = rowSums(is.na(expressionLevel_Gasch[4:176]))

max(na_count_sum)
length(na_count_sum[na_count_sum > 0 & na_count_sum < 9])
length(na_count_sum[na_count_sum > 8 & na_count_sum < 18])
length(na_count_sum[na_count_sum > 17 & na_count_sum < 44])
length(na_count_sum[na_count_sum > 43 & na_count_sum <99])
```

The 41 genes with missing values larger than 43 and smaller than 99
```{r}
names_genes = expressionLevel_Gasch$geneName
names_genes[na_count_sum > 43 & na_count_sum <99]
```



```{r echo=FALSE}
na_count_sum = rowSums(is.na(expressionLevel_Gasch[4:176]))
hist(na_count_sum, breaks = c(0,9,18,44,100), xlab="amount of missing value",ylab = "density of genes", main = 'histogram of missing value (genes)')
```

On the other hand, the distribution of missing value by 173 environmental stress indicates that there are 14 environmental stress that have a relatively higher missing value ratio (larger than 10% but smaller than 23%)

| number of environmental stresses | missing value (n)    |
| -------------------------------- | -------------------- |
| 0                                | 0                    |
| 55 (31.8%)                       | 0 < n < 62 (1%)      |
| 97 (56.1%)                       | 61 < n < 308 (5%)    |
| 7 (4.0%)                         | 307 < n < 616 (10%)  |
| 14 (8.1%)                        | 615 < n < 1385 (23%) |


```{r include=FALSE}
na_count_sum = colSums(is.na(expressionLevel_Gasch[4:176]))

max(na_count_sum)
length(na_count_sum[na_count_sum = 0])
length(na_count_sum[na_count_sum > 0 & na_count_sum < 62])
length(na_count_sum[na_count_sum > 61 & na_count_sum < 308])
length(na_count_sum[na_count_sum > 307 & na_count_sum < 616])
length(na_count_sum[na_count_sum > 615 & na_count_sum <1385])
```

```{r echo=FALSE}
hist(na_count_sum, breaks = c(0,62,308,616,1385), xlab="amount of missing value",ylab = "density of environmental stresses", main = 'histogram of missing value (environmental stresses)')
```

the 14 environmental stresses with missing value ratio larger than 10% are listed here
```{r}
na_count_sum[na_count_sum>315]
```


### <a id="filterGene"> filter out gene with high value-missing ratio (>10%) </a>
The gene with high missing value could be caused by low abundance of mRNA in the cell which means there could be large errors for the measurements on these gene. It is better to remove them first.
```{r}
na_count_sum = rowSums(is.na(expressionLevel_Gasch[4:176]))
expressionLevel_Gasch_filtered = expressionLevel_Gasch[na_count_sum<18,]
```



### <a id="imputation"> fill missing data by imputation </a>
Applying imputation to estimate the missing values should be a much better way rather than simply ignore the missing values or replacing them by mean or 0 since it reduce the bias. Olga Troyanskaya et al.(2001) suggested that weighted K-nearest neighbors (KNNimpute) could be another sensitive method to deal with the missing value.

According to their article, when we consider gene A that has one missing value under environmental stress X, we could look for how the expression level of gene A change in other environmental stress Y and looking for k other genes B that show similar behavior as gene A. Then we fill the missing value by the weighted average of genes B. The contribution of each gene is weighted by similarity of its expression to that of gene A. 

we can estimate the performance of imputation by randomly knocking out 0.1% of non-NA and check the mean square error between estimation and actual value.
```{r echo=FALSE, eval=eval_KNN_validation}

k_value_validation_df = data.frame(matrix(ncol = 2, nrow = 0))

k_list = c(2,5,8,11,14,17,20,23,27,30)
size_test = 1000

for(k in k_list){
  for(randomSeed in 362136666:362136675){
    set.seed(randomSeed)
    
    
    row_numbers = c()
    column_numbers = c()
    
    matrix = as.matrix(expressionLevel_Gasch_filtered[4:length(expressionLevel_Gasch_filtered)])
    matrix_complete = matrix
    
    set.seed(randomSeed)
    row_random = sample(1:dim(matrix)[1],size_test*2,replace = TRUE)
    column_random = sample(1:dim(matrix)[2],size_test*2,replace = TRUE)
    
    
    i = 1
    j = 1
    while (i <= round(size_test*2) & j <= size_test) {
      if(!is.na(matrix[row_random[i],column_random[i]])){
        row_numbers = c(row_numbers,row_random[i])
        column_numbers = c(column_numbers,column_random[i])
        matrix[row_random[i],column_random[i]] = NA
        j = j + 1
      }
      i = i + 1
    }
    
    
    expressionLevel_impute_knn = (impute.knn(matrix ,k = k, rowmax = 0.5, colmax = 0.8, maxp = 7000, rng.seed=362436069))$data
    
    i = 1
    se = 0
    while (i < j) {
      se = se + (as.double(matrix_complete[row_numbers[i],column_numbers[i]]) - as.double(expressionLevel_impute_knn[row_numbers[i],column_numbers[i]]))^2
      i = i + 1
    }
    
    k_value_validation_df = rbind(k_value_validation_df, c(k,se/j))
  }
}


colnames(k_value_validation_df) = c("k","mse")
k_value_validation_df$k = as.integer(k_value_validation_df$k)

k_value_validation_ds <- do.call(rbind, lapply(split(k_value_validation_df, k_value_validation_df$k), function(d) {
  data.frame(mean = mean(d$mse), sd = sd(d$mse), k = d$k)
}))

```

```{r echo=FALSE, eval=eval_KNN_validation}
ggplot() +
  geom_point(data = k_value_validation_df, aes(k, mse)) +
  geom_point(data = k_value_validation_ds, aes(k, mean), colour = 'red', size = 3) +
  geom_line(data = k_value_validation_ds, aes(k,mean), size = 1, colours = 'blue') +
  geom_errorbar(
    data = k_value_validation_ds,
    aes(k, mean, ymin = mean - sd, ymax = mean + sd),
    colour = 'red',
    width = 0.4
  )+
  theme(axis.text=element_text(size=16),axis.title=element_text(size=16,face="bold"))
```


an optimal value for k should sit between 5 and 20. In this case, we pick k = 11 for our imputation as a conserved and high-performance hyperparameter.
```{r echo=FALSE}
expressionLevel_impute_knn = (impute.knn(as.matrix(expressionLevel_Gasch_filtered[4:length(expressionLevel_Gasch_filtered)]) ,k = 11, rowmax = 0.5, colmax = 0.8, maxp = 7000, rng.seed=362436069))$data

expressionLevel_impute_knn = merge(expressionLevel_Gasch_filtered[1:3],expressionLevel_impute_knn, by = 0, sort = FALSE)
expressionLevel_impute_knn$Row.names <- NULL

fullTable_Gasch_impute_knn <- merge(expressionLevel_impute_knn,ref_motifs,by = "geneName")
```


### <a id="imputationBaseline"> baseline performance: simply replace missing value with mean of column or 0 </a>
##### NA to column mean
As one of the easiest approach to deal with the missing values, we could simply replace them by mean of that column
To estimate of the performance, we randomly knock out 0.1% of non-NA and check the mean square error.
```{r echo=FALSE}
baseline_df = data.frame(matrix(ncol = 1, nrow = 0))
expressionLevel_matrix = as.matrix(expressionLevel_Gasch_filtered[4:length(expressionLevel_Gasch_filtered)])
storage.mode(expressionLevel_matrix) <- "double"
expressionLevel_column_mean = colMeans(expressionLevel_matrix,na.rm = TRUE)

for(randomSeed in 362136666:362136675){
  size_test = 1000
  row_numbers = c()
  column_numbers = c()
  
  
  set.seed(randomSeed)
  row_random = sample(1:dim(expressionLevel_matrix)[1],size_test*2,replace = TRUE)
  column_random = sample(1:dim(expressionLevel_matrix)[2],size_test*2,replace = TRUE)
  
  i = 1
  j = 1
  while (i <= round(size_test*2) & j <= size_test) {
    if(!is.na(expressionLevel_matrix[row_random[i],column_random[i]])){
      row_numbers = c(row_numbers,row_random[i])
      column_numbers = c(column_numbers,column_random[i])
      j = j + 1
    }
    i = i + 1
  }
  
  
  se = 0
  for (i in 1:(j-1)) {
    se = se + (expressionLevel_matrix[row_numbers[i],column_numbers[i]] - expressionLevel_column_mean[column_numbers[i]])^2
  }
  
  baseline_df = rbind(baseline_df, c(se/j))
  print(paste("estimate mse = ",se/j))
  
}

colnames(baseline_df) = c("mse")

```

```{r echo=FALSE}
baseline_ds = data.frame(matrix(ncol = 2, nrow = 0))
baseline_ds = rbind(baseline_ds, c(mean(baseline_df$mse),sd(baseline_df$mse)))
colnames(baseline_ds) = c("mean","sd")

```

##### NA to zero
```{r echo=FALSE}
baseline2_df = data.frame(matrix(ncol = 1, nrow = 0))
expressionLevel_matrix = as.matrix(expressionLevel_Gasch_filtered[4:length(expressionLevel_Gasch_filtered)])
storage.mode(expressionLevel_matrix) <- "double"
expressionLevel_column_mean = colMeans(expressionLevel_matrix,na.rm = TRUE)

for(randomSeed in 362136666:362136675){
  size_test = 1000
  se = 0
  
  set.seed(randomSeed)
  row_random = sample(1:dim(expressionLevel_matrix)[1],size_test*2,replace = TRUE)
  column_random = sample(1:dim(expressionLevel_matrix)[2],size_test*2,replace = TRUE)
  
  i = 1
  j = 1
  while (i <= round(size_test*2) & j <= size_test) {
    if(!is.na(expressionLevel_matrix[row_random[i],column_random[i]])){
      se = se + (expressionLevel_matrix[row_random[i],column_random[i]])^2
      j = j + 1
    }
    i = i + 1
  }
  
  baseline2_df = rbind(baseline2_df, c(se/j))
  print(paste("estimate mse = ",se/j))
}

colnames(baseline2_df) = c("mse")

```

```{r echo=FALSE}
baseline2_ds = data.frame(matrix(ncol = 2, nrow = 0))
baseline2_ds = rbind(baseline2_ds, c(mean(baseline2_df$mse),sd(baseline2_df$mse)))
colnames(baseline2_ds) = c("mean","sd")
```

##### compare the between performance of kNN imputation, NA to column mean and NA to zero
```{r eval=eval_KNN_validation, fig.width=8}
ggplot() +
  geom_point(data = baseline_df, aes("NA to column mean", mse)) +
  geom_point(data = baseline2_df, aes("NA to zero", mse)) +
  geom_point(data = k_value_validation_df[k_value_validation_df$k == 11,], aes("kNN imputation (k=11)",mse)) +
  geom_point(data = baseline_ds, aes("NA to column mean", mean), colour = 'red', size = 3) +
  geom_point(data = baseline2_ds, aes("NA to zero", mean), colour = 'red', size = 3) +
  geom_point(data = k_value_validation_ds[k_value_validation_ds$k == 11,], aes("kNN imputation (k=11)", mean), colour = 'red', size = 3) +
  geom_errorbar(
    data = baseline_ds,
    aes("NA to column mean", mean, ymin = mean - sd, ymax = mean + sd),
    colour = 'red',
    width = 0.2
  ) + 
  geom_errorbar(
    data = baseline2_ds,
    aes("NA to zero", mean, ymin = mean - sd, ymax = mean + sd),
    colour = 'red',
    width = 0.2
  ) +
  geom_errorbar(
    data = k_value_validation_ds[k_value_validation_ds$k == 11,],
    aes("kNN imputation (k=11)", mean, ymin = mean - sd, ymax = mean + sd),
    colour = 'red',
    width = 0.2
  ) +
  theme(axis.text=element_text(size=16),axis.title=element_text(size=16,face="bold")) +
  labs(x = "")+
  ylim(0,0.8)

```


### consider only motifs with non-zero frequency
```{r}
motifs_count_sum <- colSums(fullTable_Gasch[names_motifs_all],na.rm=TRUE)
remove_list = NULL

for (i in names(motifs_count_sum)){
  if (motifs_count_sum[[i]] == 0){
    remove_list <- c(remove_list,i)
  }
}

names_motifs_valid <- names_motifs_all[!names_motifs_all %in% remove_list]
```

### <a id="crossValidation"> choose lambda from 10-fold cross validation </a>
We need to set the hyperparameter lambda to control the effect of penalization. A general approch to this is by cross validation.
```{r echo=FALSE}
x = as.matrix(fullTable_Gasch_impute_knn[names_motifs_valid])
y = as.matrix(fullTable_Gasch_impute_knn[names_environmental_stress])
```

```{r eval=eval_crossValidation}
cv_models_all = cv.glmnet(x, y, family = "mgaussian")
plot(cv_models_all)
```

`lambda.min` is the value of lambda that gives minimum cvm; `lambda.1se` is the largest value of lambda such that error is within 1 standard error of the minimum; The confidence intervals represent error estimates for the loss metric (red dots). They're computed using cross validation. The vertical dashed lines show the locations of λmin and λ1se. The numbers across the top are the number of nonzero coefficient estimates. The error is accumulated, and the average error and standard deviation over the folds is computed. 

### <a id="trainingModels"> training model with selected lambda </a>
training model with λ = `lambda.min` results in a model with highest expected performancec while training model with λ = `lambda.1se` results in a more conservative model with fewer non-zero regression coefficients.

```{r eval=!eval_crossValidation}
# setting lambda manually when eval_crossValidation = False
lambda.min = 0.0334379510944218
lambda.1se = 0.235898134217399

models_all_minLambda = glmnet(x, y, family = "mgaussian", lambda = lambda.min)
models_all_1seLambda = glmnet(x, y, family = "mgaussian", lambda = lambda.1se)
```

```{r eval=eval_crossValidation}
models_all_minLambda = glmnet(x, y, family = "mgaussian", lambda = cv_models_all$lambda.min)
models_all_1seLambda = glmnet(x, y, family = "mgaussian", lambda = cv_models_all$lambda.1se)

paste("lambda.min =",cv_models_all$lambda.min)
paste("lambda.1se =",cv_models_all$lambda.1se)
```


### the penalize effect with λ = lambda.1se is much stronger than that with λ = lambda.min
We summary coefficients from all the models together to compare the penalized effect of models with different λ values. The model with λ = lambda.1se penalizes about 80% of coefficients to 0 while the model with λ = lambda.min keep more than half of its coefficients non-zero.

```{r echo=FALSE}
summary_lambda_coefficient = data.frame(matrix(ncol = 3, nrow = 0))

{
  count_nonzero_coefficient = 0
  for (condition in names_environmental_stress) {
    count_nonzero_coefficient = count_nonzero_coefficient + models_all_minLambda$beta[[condition]]@p[2]
  }
  
  count_all_coefficient = length(names_environmental_stress)*length(names_motifs_valid)
}

summary_lambda_coefficient = rbind(summary_lambda_coefficient, data.frame(
    lambda = "min",
    coefficient = "non-zero",
    percentage = round(count_nonzero_coefficient/count_all_coefficient,2)
  ))
summary_lambda_coefficient = rbind(summary_lambda_coefficient, data.frame(
    lambda = "min",
    coefficient = "zero",
    percentage = (1 - round(count_nonzero_coefficient/count_all_coefficient,2))
  ))


{
  count_nonzero_coefficient = 0
  for (condition in names_environmental_stress) {
    count_nonzero_coefficient = count_nonzero_coefficient + models_all_1seLambda$beta[[condition]]@p[2]
  }
  
  count_all_coefficient = length(names_environmental_stress)*length(names_motifs_valid)
}

summary_lambda_coefficient = rbind(summary_lambda_coefficient, data.frame(
    lambda = "1se",
    coefficient = "non-zero",
    percentage = round(count_nonzero_coefficient/count_all_coefficient,2)
  ))
summary_lambda_coefficient = rbind(summary_lambda_coefficient, data.frame(
    lambda = "1se",
    coefficient = "zero",
    percentage = (1 - round(count_nonzero_coefficient/count_all_coefficient,2))
  ))
```


```{r echo=FALSE}
ggplot(data=summary_lambda_coefficient, aes(x=lambda, y=percentage, fill=coefficient), labdels = percentage) +
  geom_bar(stat="identity", position=position_dodge())+
  geom_text(aes(label=percentage), position=position_dodge(width=0.9), vjust=-0.25)+
  scale_fill_brewer(palette="Paired")+
  theme(axis.text.x = element_text(size = 14),
        axis.text.y = element_text(size = 14),
        axis.title = element_text(size = 14))+
  ylab("percentage of coefficient")
  
```

```{r echo=FALSE}
coefficients = NULL
for (condition in names_environmental_stress) {
  coefficients = c(coefficients,models_all_minLambda$beta[[condition]]@x)
}

hist(coefficients, main = 'histgram of non-zero coefficient when λ = lambda.min')
```


```{r echo=FALSE}
coefficients = NULL
for (condition in names_environmental_stress) {
  coefficients = c(coefficients,models_all_1seLambda$beta[[condition]]@x)
}

hist(coefficients, main = 'histgram of non-zero coefficient when λ = lambda.1se')

```



### regulatory effect of motifs peaks at different times and decrease as time pass
By extracting the regression coefficients given to motifs and sorting them with the previous grouping, we found that the regression coefficients of most motifs show a tendency to rise first, peak in a certain period and then fall. It describes the process of the regulation effect which gradually increases and returns to a stable state after the environmental stress. The time when the regulation effect reaches its peak varies with the type of environmental stress. For example, undereat shock-related environmental stress, the regression coefficients of most motifs peaked in about 15 minutes. while under diamide treatment, the regression coefficient of motifs generally reaches its peak after 30 minutes. It is worth mentioning that the degree of environmental stress has no obvious effect on the peaking period. As shown in Figure 5-1 and Figure 5-3, for most of the motifs, both their regression coefficients in `heat shock from 25°C to 37° C` and `heat shock from 29°C to 33° C `peaked at around 15 minutes. On the other hand, even under the same environmental conditions, not all the regression coefficients peak at the same period. This may because of the complex regulatory mechanism within the cell. While the cell adjusts its gene expression level, the number of regulatory elements (as products of the gene expression) such as RNA binding proteins that participate in regulation also changes, and lead to a change in the regulatory effect of motifs change as well.

```{r fig.height=7, fig.width=9}

names_temperature_condition = c(
  "hs_05min_hs-1",
  "hs_10min_hs-1",
  "hs_15min_hs-1",
  "hs_30min_hs-1",
  "hs_40min_hs-1",
  "hs_60min_hs-1",
  "hs_80min_hs-1"
)

plot_models_line(models_all_minLambda, names_temperature_condition, 10)
```

```{r fig.height=7, fig.width=9}

names_temperature_condition = c(
  "hs_29to33_05min",
  "hs_29to33_15min",
  "hs_29to33_30min"
)

plot_models_line(models_all_minLambda, names_temperature_condition, 10)
```


```{r fig.height=7, fig.width=9}

names_diamide_condition = c(
  "1.5mM_diamide_5min",
  "1.5mM_diamide_10min",
  "1.5mM_diamide_20min",
  "1.5mM_diamide_30min",
  "1.5mM_diamide_40min",
  "1.5mM_diamide_50min",
  "1.5mM_diamide_60min",
  "1.5mM_diamide_90min"
)

plot_models_line(models_all_minLambda, names_diamide_condition, 10)
```





### select the period when gene expression effect is biggest with variance of log2 expression level


| type of environmental stress | period when gene expression effect is biggest (min) |
| :--------------------------: | :-------------------------------------------------: |
|          heatshock           |                       15 ~ 20                       |
|         hypothermia          |                       15, 60                        |
|      sorbitol treatment      |                       30 ~ 45                       |
|      nitrogen depletion      |                     2880 ~ 7200                     |
|      diamide treatment       |                       30 ~ 40                       |
|   dithiothrietol exposure    |                       20 ~ 60                       |
|      menadione exposure      |                  30 ~ 40,  80~120                   |
| hydrogen peroxide treatment  |                       20 ~ 30                       |


```{r}
expressionLevel_matrix = as.matrix(expressionLevel_impute_knn[4:length(expressionLevel_Gasch)])
storage.mode(expressionLevel_matrix) <- "double"
expressionLevel_column_variance = colMeans(expressionLevel_matrix*expressionLevel_matrix)-colMeans(expressionLevel_matrix)*colMeans(expressionLevel_matrix)
```

```{r}
expressionLevel_column_variance
```




# Part 4: analyze the regression coefficient
### <a id="environmentalGroup"> divide environmental condition into groups </a>
With the models from group lasso, we can carefully select the regression coefficient from different groups of environmental conditions and compare them. By calculating the sum and L2 norm of their coefficient under each group of environmental condition, we are able to compare how the regulatory effect of each motif differ in each environmental stress. 

```{r}
names_heatshock_29to33_condition = c(
  "hs_29to33_05min",
  "hs_29to33_15min",
  "hs_29to33_30min"
)

names_heatshock_sorbitol_condition = c(
  "29C(1M_sorbitol)~33C(1M_sorbitol)_05min",
  "29C(1M_sorbitol)~33C(1M_sorbitol)_15min",
  "29C(1M_sorbitol)~33C(1M_sorbitol)_30min"
)

names_hypothermia_condition = c(
  "hs_37to25_15min",
  "hs_37to25_30min",
  "hs_37to25_60min"
)

names_heatshock_25to37_condition = c(
  "hs_15min_hs-1",
  "hs_30min_hs-1",
  "hs_60min_hs-1"
)

```



```{r echo=FALSE}
# prepare summary dataframe
summary_coefficient = data.frame(matrix(ncol = 6, nrow = 0))

coefficients = extract_coefficient(models_all_minLambda,names_environmental_stress)

for(motif in names_motifs_valid){
  coefficients_motif_specific = coefficients[coefficients$rn == motif,]$value
  summary_coefficient = rbind(summary_coefficient, data.frame(
    motif = motif,
    condition = "all",
    sum = sum(coefficients_motif_specific),
    mean = sum(coefficients_motif_specific)/length(names_environmental_stress),
    L2 = sqrt(sum(coefficients_motif_specific^2)),
    meanL2 = sqrt(sum(coefficients_motif_specific^2)/length(names_environmental_stress))
  ))
}




coefficients = extract_coefficient(models_all_minLambda,names_heatshock_29to33_condition)

for(motif in names_motifs_valid){
  coefficients_motif_specific = coefficients[coefficients$rn == motif,]$value
  summary_coefficient = rbind(summary_coefficient, data.frame(
    motif = motif,
    condition = "heat shock from 29C to 33C",
    sum = sum(coefficients_motif_specific),
    mean = sum(coefficients_motif_specific)/length(names_heatshock_29to33_condition),
    L2 = sqrt(sum(coefficients_motif_specific^2)),
    meanL2 = sqrt(sum(coefficients_motif_specific^2)/length(names_heatshock_29to33_condition))
  ))
}


coefficients = extract_coefficient(models_all_minLambda,names_heatshock_sorbitol_condition)

for(motif in names_motifs_valid){
  coefficients_motif_specific = coefficients[coefficients$rn == motif,]$value
  summary_coefficient = rbind(summary_coefficient, data.frame(
    motif = motif,
    condition = "heat shock with sorbitol",
    sum = sum(coefficients_motif_specific),
    mean = sum(coefficients_motif_specific)/length(names_heatshock_sorbitol_condition),
    L2 = sqrt(sum(coefficients_motif_specific^2)),
    meanL2 = sqrt(sum(coefficients_motif_specific^2)/length(names_heatshock_sorbitol_condition))
  ))
}


coefficients = extract_coefficient(models_all_minLambda,names_hypothermia_condition)

for(motif in names_motifs_valid){
  coefficients_motif_specific = coefficients[coefficients$rn == motif,]$value
  summary_coefficient = rbind(summary_coefficient, data.frame(
    motif = motif,
    condition = "hypothermia from 37C to 25C",
    sum = sum(coefficients_motif_specific),
    mean = sum(coefficients_motif_specific)/length(names_hypothermia_condition),
    L2 = sqrt(sum(coefficients_motif_specific^2)),
    meanL2 = sqrt(sum(coefficients_motif_specific^2)/length(names_hypothermia_condition))
  ))
}


coefficients = extract_coefficient(models_all_minLambda,names_heatshock_25to37_condition)

for(motif in names_motifs_valid){
  coefficients_motif_specific = coefficients[coefficients$rn == motif,]$value
  summary_coefficient = rbind(summary_coefficient, data.frame(
    motif = motif,
    condition = "heat shock from 25C to 37C",
    sum = sum(coefficients_motif_specific),
    mean = sum(coefficients_motif_specific)/length(names_heatshock_25to37_condition),
    L2 = sqrt(sum(coefficients_motif_specific^2)),
    meanL2 = sqrt(sum(coefficients_motif_specific^2)/length(names_heatshock_25to37_condition))
  ))
}

```

```{r}
summary_motif_frequency = data.frame(matrix(ncol= 2, nrow = 0))

for(motif in names_motifs_valid){
  summary_motif_frequency = rbind(summary_motif_frequency, data.frame(
    motif = motif,
    transcriptAmount = log10(sum(fullTable_Gasch[[motif]] > 0))
  ))
}

```

```{r}
summary_coefficient = merge(x = summary_coefficient, y = summary_motif_frequency, by = "motif")
```



### <a id = "L2Allcondition"> L2 norm of the non-zero regression coefficient from significant motif under all condition </a>
The following graph shows the L2 norm of the regression coefficient of some significant motif under all environmental conditions. The rank of each motif shows how significant a motif is in the environmental conditions we considered. However, since the models we fitted here only considered some typical environmental stress on yeast and is very unbalanced in the type of condition (about 40 model is temperature relevant), we couldn't say the high-rank motif shown here is significant under general environmental stress.

```{r echo=FALSE, fig.height=8, fig.width=14}
temp_summary_coefficient = summary_coefficient[summary_coefficient$condition == "all",]
temp_summary_coefficient = temp_summary_coefficient[order(-temp_summary_coefficient$L2),]


ggplot(data=head(temp_summary_coefficient,31), aes(x=reorder(motif,-L2), y=L2, fill=condition)) +
  geom_bar(stat="identity", position=position_dodge()) +
  theme(axis.text.x = element_text(size = 16, angle = 70, hjust = 1),
        axis.text.y = element_text(size = 16),
        legend.text = element_text(size = 16),
        axis.title = element_text(size = 16),
        title = element_text(size = 14)) + 
  scale_x_discrete(name ="Motif") + 
  ggtitle("L2 of the non-zero regression coefficient from significant motif under all condition")
```

### <a id = "heatShockHypothermia"> compare L2 nore and sum of the non-zero regression coefficient from significant motif under heat shock and hypothermia </a>
the regulatory effect of each motif under hypothermia is much smaller than that under heat shock. This could probably reflect the trigger of environmental stress response (ESR) mention by Gasch et al. (2000): "the ESR is not initiated in response to all environmental changes and that the large, transient changes in expression that are characteristic of the ESR are only seen when this response is initiated and not in the reciprocal response to diminished environmental stress."

```{r echo=FALSE, fig.height=8, fig.width=14}
temp_summary_coefficient = summary_coefficient[summary_coefficient$condition == "heat shock from 25C to 37C" | summary_coefficient$condition == "hypothermia from 37C to 25C",]
temp_summary_coefficient = temp_summary_coefficient[order(-temp_summary_coefficient$L2),]
temp_summary_coefficient = temp_summary_coefficient[temp_summary_coefficient$motif %in% head(unique(temp_summary_coefficient$motif),20),]


ggplot(data=temp_summary_coefficient, aes(x=reorder(motif,-L2), y=meanL2, fill=condition)) +
  geom_bar(stat="identity", position=position_dodge()) +
  theme(axis.text.x = element_text(size = 16, angle = 70, hjust = 1),
        axis.text.y = element_text(size = 16),
        legend.text = element_text(size = 16),
        axis.title = element_text(size = 16)) + 
  scale_x_discrete(name ="Motif") + 
  ggtitle("mean L2 of the non-zero regression coefficient of significant motifs under heat shock and hypothermia after 15, 30 and 60 min")
```

```{r echo=FALSE, fig.height=8, fig.width=14}
temp_summary_coefficient = summary_coefficient[summary_coefficient$condition == "heat shock from 25C to 37C" | summary_coefficient$condition == "hypothermia from 37C to 25C",]
temp_summary_coefficient = temp_summary_coefficient[order(-temp_summary_coefficient$L2),]
temp_summary_coefficient = temp_summary_coefficient[temp_summary_coefficient$motif %in% head(unique(temp_summary_coefficient$motif),20),]


ggplot(data=temp_summary_coefficient, aes(x=reorder(motif,-L2), y=mean, fill=condition)) +
  geom_bar(stat="identity", position=position_dodge()) +
  theme(axis.text.x = element_text(size = 16, angle = 70, hjust = 1),
        axis.text.y = element_text(size = 16),
        legend.text = element_text(size = 16),
        axis.title = element_text(size = 16)) + 
  scale_x_discrete(name ="Motif") + 
  ggtitle("Mean of the non-zero regression coefficient of significant motifs under heat shock and hypothermia after 15, 30 and 60 min")
```


### compare L2 norm and sum of the non-zero regression coefficient from significant motif under heat shock, heat shock with sorbitol
Although most of the regression coefficients are similar in both heat shock with and without sorbitol, there are some motifs show a relatively different regulatory effect on gene expression with sorbitol. Check the second graph `Sum of the non-zero regression coefficient of significant motifs under heat shock with and without sorbitol after 5, 15 and 30 min`, it is easy to see `TGAGGCTA` shows a much stronger regulatory effect on gene expression when there is sorbitol while `UKWCGRGGN` and `GTAGTATTT` acting in an opposite way, a much weaker regulatory effect under heat shock with sorbitol.


```{r echo=FALSE, fig.height=8, fig.width=14}
temp_summary_coefficient = summary_coefficient[summary_coefficient$condition == "heat shock from 29C to 33C" | summary_coefficient$condition == "heat shock with sorbitol",]
temp_summary_coefficient = temp_summary_coefficient[order(-temp_summary_coefficient$L2),]
temp_summary_coefficient = temp_summary_coefficient[temp_summary_coefficient$motif %in% head(unique(temp_summary_coefficient$motif),20),]



ggplot(data=temp_summary_coefficient, aes(x=reorder(motif,-L2), y=L2, fill=condition)) +
  geom_bar(stat="identity", position=position_dodge()) +
  theme(axis.text.x = element_text(size = 16, angle = 70, hjust = 1),
        axis.text.y = element_text(size = 16),
        legend.text = element_text(size = 16),
        axis.title = element_text(size = 16),
        title = element_text(size = 14)) + 
  scale_x_discrete(name ="Motif") + 
  ggtitle("L2 of the non-zero regression coefficient of significant motifs under heat shock with and without sorbitol after 5, 15 and 30 min")
```



```{r echo=FALSE, fig.height=8, fig.width=14}
temp_summary_coefficient = summary_coefficient[summary_coefficient$condition == "heat shock from 29C to 33C" | summary_coefficient$condition == "heat shock with sorbitol",]
temp_summary_coefficient = temp_summary_coefficient[order(-temp_summary_coefficient$L2),]
temp_summary_coefficient = temp_summary_coefficient[temp_summary_coefficient$motif %in% head(unique(temp_summary_coefficient$motif),20),]


ggplot(data=temp_summary_coefficient, aes(x=reorder(motif,-L2), y=sum, fill=condition)) +
  geom_bar(stat="identity", position=position_dodge()) +
  theme(axis.text.x = element_text(size = 16, angle = 70, hjust = 1),
        axis.text.y = element_text(size = 16),
        legend.text = element_text(size = 16),
        axis.title = element_text(size = 16)) + 
  scale_x_discrete(name ="Motif") + 
  ggtitle("Sum of the non-zero regression coefficient of significant motifs under heat shock with and without sorbitol after 5, 15 and 30 min")
```




### <a id="frequencyCoefficient">investigate the transcript frequency of each motif against their coefficient</a>

```{r}
order_motifs = summary_coefficient[summary_coefficient$condition == "all",]
order_motifs = order_motifs[order(-order_motifs$transcriptAmount),]$motif
```


```{r echo=FALSE, fig.height=8, fig.width=14}
temp_summary_coefficient = summary_coefficient[summary_coefficient$condition == "all",]
temp_summary_coefficient$motif = factor(temp_summary_coefficient$motif, levels = order_motifs)
temp_summary_coefficient = temp_summary_coefficient[temp_summary_coefficient$motif %in% head(order_motifs,31),]


ggplot(data=temp_summary_coefficient, aes(x=motif)) +
  geom_bar(stat="identity", aes(y=meanL2,fill = condition)) +
  geom_line(aes(y = transcriptAmount/2.5, colour = "log10 transcript frequency"),group=1, size=2) +
  geom_text(aes(y = transcriptAmount/2.5, label = round(transcriptAmount, 1)), vjust = -1.0, color = "black", size = 4) +
  theme(axis.text.x = element_text(size = 16, angle = 70, hjust = 1),
        axis.text.y = element_text(size = 16),
        legend.text = element_text(size = 16),
        axis.title = element_text(size = 16)) + 
  scale_colour_manual(name="", values=c("log10 transcript frequency"="gray20"))+
  scale_x_discrete(name ="Motif") + 
  ylim(0,1.5)+
  ggtitle("frequency vs mean L2 of each motif under all condition")
```


```{r echo=FALSE, fig.height=8, fig.width=14}
temp_summary_coefficient = summary_coefficient[summary_coefficient$condition == "all",]
temp_summary_coefficient$motif = factor(temp_summary_coefficient$motif, levels = order_motifs)
temp_summary_coefficient = temp_summary_coefficient[temp_summary_coefficient$motif %in% tail(order_motifs,31),]


ggplot(data=temp_summary_coefficient, aes(x=motif)) +
  geom_bar(stat="identity", aes(y=meanL2,fill = condition)) +
  geom_line(aes(y = transcriptAmount/2.5, colour = "log10 transcript frequency"),group=1, size=2) +
  geom_text(aes(y = transcriptAmount/2.5, label = round(transcriptAmount, 1)), vjust = -1.0, color = "black", size = 4) +
  theme(axis.text.x = element_text(size = 16, angle = 70, hjust = 1),
        axis.text.y = element_text(size = 16),
        legend.text = element_text(size = 16),
        axis.title = element_text(size = 16)) + 
  scale_colour_manual(name="", values=c("log10 transcript frequency"="gray20"))+
  scale_x_discrete(name ="Motif") + 
  ylim(0,1.5)+
  ggtitle("frequency vs L2 of each motif under all condition")
```


```{r echo=FALSE, fig.height=8, fig.width=14}
temp_summary_coefficient = summary_coefficient[summary_coefficient$condition == "heat shock from 29C to 33C" | summary_coefficient$condition == "heat shock with sorbitol",]
temp_summary_coefficient$motif = factor(temp_summary_coefficient$motif, levels = order_motifs)
temp_summary_coefficient = temp_summary_coefficient[temp_summary_coefficient$motif %in% head(order_motifs,31),]


ggplot(data=temp_summary_coefficient, aes(x=motif)) +
  geom_bar(stat="identity", position=position_dodge(), aes(y=meanL2,fill = condition)) +
  geom_line(aes(y = transcriptAmount/3, colour = "log10 transcript frequency"),group=1, size=2) +
  geom_text(aes(y = transcriptAmount/3, label = round(transcriptAmount, 1)), vjust = -1.0, color = "black", size = 4) +
  theme(axis.text.x = element_text(size = 16, angle = 70, hjust = 1),
        axis.text.y = element_text(size = 16),
        legend.text = element_text(size = 16),
        axis.title = element_text(size = 16)) + 
  scale_colour_manual(name="", values=c("log10 transcript frequency"="gray20"))+
  scale_x_discrete(name ="Motif") + 
  ylim(0,1.2)+
  ggtitle("frequency vs mean L2 of each motif under heat shock with and without sorbitol")
```



```{r echo=FALSE, fig.height=8, fig.width=14}
temp_summary_coefficient = summary_coefficient[summary_coefficient$condition == "heat shock from 29C to 33C" | summary_coefficient$condition == "heat shock with sorbitol",]
temp_summary_coefficient$motif = factor(temp_summary_coefficient$motif, levels = order_motifs)
temp_summary_coefficient = temp_summary_coefficient[temp_summary_coefficient$motif %in% tail(order_motifs,31),]


ggplot(data=temp_summary_coefficient, aes(x=motif)) +
  geom_bar(stat="identity", position=position_dodge(), aes(y=meanL2,fill = condition)) +
  geom_line(aes(y = transcriptAmount/3, colour = "log10 transcript frequency"),group=1, size=2) +
  geom_text(aes(y = transcriptAmount/3, label = round(transcriptAmount, 1)), vjust = -1.0, color = "black", size = 4) +
  theme(axis.text.x = element_text(size = 16, angle = 70, hjust = 1),
        axis.text.y = element_text(size = 16),
        legend.text = element_text(size = 16),
        axis.title = element_text(size = 16)) + 
  scale_colour_manual(name="", values=c("log10 transcript frequency"="gray20"))+
  scale_x_discrete(name ="Motif") + 
  ylim(0,1.2)+
  ggtitle("frequency vs L2 of each motif under all condition")
```




### <a id="boxPlot"> box graph for cross comparison between two type of environmental stress </a>
To compare how the coefficient of motif from two different type of environmental stress, we plot a box graph show the difference between the coefficients from two group of models.
```{r}
names_heatshock_29to33_condition = c(
  "hs_29to33_05min",
  "hs_29to33_15min",
  "hs_29to33_30min"
)

names_heatshock_sorbitol_condition = c(
  "29C(1M_sorbitol)~33C(1M_sorbitol)_05min",
  "29C(1M_sorbitol)~33C(1M_sorbitol)_15min",
  "29C(1M_sorbitol)~33C(1M_sorbitol)_30min"
)

```


```{r echo=FALSE}
temp_coefficients_1 = extract_coefficient(models_all_minLambda, names_heatshock_29to33_condition)
temp_coefficients_1$condition = rep("heat shock without sorbitol",nrow(temp_coefficients_1))

temp_coefficients_2 = extract_coefficient(models_all_minLambda, names_heatshock_sorbitol_condition)
temp_coefficients_2$condition = rep("heat shock with sorbitol",nrow(temp_coefficients_2))

coefficients = rbind(temp_coefficients_1,temp_coefficients_2)
colnames(coefficients) = c("motif","variable","coefficient","condition")
```


```{r echo=FALSE, fig.height=8, fig.width=14}
coefficients$motif = factor(coefficients$motif, levels = order_motifs)

ggplot(data=coefficients[coefficients$motif %in% head(order_motifs,31)], aes(x=motif,y=coefficient)) +
  geom_boxplot() +
  geom_point(aes(colour=condition), size = 3) +
  theme(axis.text.x = element_text(size = 16, angle = 70, hjust = 1),
        axis.text.y = element_text(size = 16),
        legend.text = element_text(size = 16),
        axis.title = element_text(size = 16)) + 
  scale_x_discrete(name ="Motif") + 
  ggtitle("coefficients from models of heat shock with and without sorbitol")
```


```{r echo=FALSE, fig.height=8, fig.width=14}
ggplot(data=coefficients[coefficients$motif %in% tail(order_motifs,31)], aes(x=motif,y=coefficient)) +
  geom_boxplot() +
  geom_point(aes(colour=condition), size = 3) +
  theme(axis.text.x = element_text(size = 16, angle = 70, hjust = 1),
        axis.text.y = element_text(size = 16),
        legend.text = element_text(size = 16),
        axis.title = element_text(size = 16)) + 
  scale_x_discrete(name ="Motif") + 
  ggtitle("coefficients from models of heat shock with and without sorbitol")
```



```{r}

names_heatshock_25to37_condition = c(
  "hs_05min_hs-1",
  "hs_10min_hs-1",
  "hs_15min_hs-1",
  "hs_20min_hs-1",
  "hs_30min_hs-1",
  "hs_40min_hs-1",
  "hs_60min_hs-1",
  "hs_80min_hs-1"
)

names_heatshock_37to25_condition = c(
  "hs_37to25_15min",
  "hs_37to25_30min",
  "hs_37to25_45min",
  "hs_37to25_60min",
  "hs_37to25_90min"
)

```


```{r echo=FALSE}
temp_coefficients_1 = extract_coefficient(models_all_minLambda, names_heatshock_25to37_condition)
temp_coefficients_1$condition = rep("heat shock",nrow(temp_coefficients_1))

temp_coefficients_2 = extract_coefficient(models_all_minLambda, names_heatshock_37to25_condition)
temp_coefficients_2$condition = rep("hypothermia",nrow(temp_coefficients_2))

coefficients = rbind(temp_coefficients_1,temp_coefficients_2)
colnames(coefficients) = c("motif","variable","coefficient","condition")
```


```{r echo=FALSE, fig.height=8, fig.width=14}
coefficients$motif = factor(coefficients$motif, levels = order_motifs)

ggplot(data=coefficients[coefficients$motif %in% head(order_motifs,31)], aes(x=motif,y=coefficient)) +
  geom_boxplot() +
  geom_point(aes(colour=condition), size = 3) +
  theme(axis.text.x = element_text(size = 16, angle = 70, hjust = 1),
        axis.text.y = element_text(size = 16),
        legend.text = element_text(size = 16),
        axis.title = element_text(size = 16)) + 
  scale_x_discrete(name ="Motif") + 
  ggtitle("coefficients from heat shock and hypothermia models")
```


```{r echo=FALSE, fig.height=8, fig.width=14}
ggplot(data=coefficients[coefficients$motif %in% tail(order_motifs,31)], aes(x=motif,y=coefficient)) +
  geom_boxplot() +
  geom_point(aes(colour=condition), size = 3) +
  theme(axis.text.x = element_text(size = 16, angle = 70, hjust = 1),
        axis.text.y = element_text(size = 16),
        legend.text = element_text(size = 16),
        axis.title = element_text(size = 16)) + 
  scale_x_discrete(name ="Motif") + 
  ggtitle("coefficients from heat shock and hypothermia models")
```




### <a id="motifRank"> comparison motifs rank between each group of models </a>
As another approach from comparing the motifs' regulatory effect between different type of environmental stresses, we compare the absolute value of their regression coefficients. Motifs are ranked in each model where rank 1 is given to the motifs with highest absolute value of coefficient. Then, for each group of environmental stress, we compute the average rank of each motif weighted by L2 of the coefficients in each model. Therefore, the rank from a model with motifs with low regression coefficients will get less weight rather than that of a model with motifs with high regression coefficients.
```{r}
# prepare data
summary_coefficient_rank = data.frame(matrix(ncol = 4, nrow = 0))

for(environmental_stress in names_environmental_stress){
  coefficients = extract_coefficient(models_all_minLambda,environmental_stress)
  #coefficients = coefficients[order(abs(coefficients$value)),]
  coefficients = coefficients[order(-abs(coefficients$value)),]

  i = 1
  while (i <= dim(coefficients)[1]) {
    summary_coefficient_rank = rbind(summary_coefficient_rank, data.frame(
    motif = coefficients[i]$rn,
    rank = i,
    condition = environmental_stress,
    model.L2 = sqrt(sum(coefficients$value^2))
    ))
    i = i+1
  }
}

```


```{r}
{
  names_heatshock_25to37_condition = vector(mode="character", length=8)
  names_heatshock_25to37_condition[1] = "hs_05min_hs-1"
  names_heatshock_25to37_condition[2] = "hs_10min_hs-1"
  names_heatshock_25to37_condition[3] = "hs_15min_hs-1"
  names_heatshock_25to37_condition[4] = "hs_20min_hs-1"
  names_heatshock_25to37_condition[5] = "hs_30min_hs-1"
  names_heatshock_25to37_condition[6] = "hs_40min_hs-1"
  names_heatshock_25to37_condition[7] = "hs_60min_hs-1"
  names_heatshock_25to37_condition[8] = "hs_80min_hs-1"
}

{
  names_heatshock_37to25_condition = vector(mode="character", length=5)
  names_heatshock_37to25_condition[1] = "hs_37to25_15min"
  names_heatshock_37to25_condition[2] = "hs_37to25_30min"
  names_heatshock_37to25_condition[3] = "hs_37to25_45min"
  names_heatshock_37to25_condition[4] = "hs_37to25_60min"
  names_heatshock_37to25_condition[5] = "hs_37to25_90min"
}

{
  names_heatshock_sorbitol_condition = vector(mode="character", length=3)
  names_heatshock_sorbitol_condition[1] = "29C(1M_sorbitol)~33C(1M_sorbitol)_05min"
  names_heatshock_sorbitol_condition[2] = "29C(1M_sorbitol)~33C(1M_sorbitol)_15min"
  names_heatshock_sorbitol_condition[3] = "29C(1M_sorbitol)~33C(1M_sorbitol)_30min"
}

{
  names_sorbitol_condition = vector(mode="character", length=7)
  names_sorbitol_condition[1] = "1M_sorbitol_05min"
  names_sorbitol_condition[2] = "1M_sorbitol_15min"
  names_sorbitol_condition[3] = "1M_sorbitol_30min"
  names_sorbitol_condition[4] = "1M_sorbitol_45min"
  names_sorbitol_condition[5] = "1M_sorbitol_60min"
  names_sorbitol_condition[6] = "1M_sorbitol_90min"
  names_sorbitol_condition[7] = "1M_sorbitol_120min"
}

{
  names_Nitrogen_condition = vector(mode="character", length=7)
  names_Nitrogen_condition[1] = "Nitrogen_Depletion_30min"
  names_Nitrogen_condition[2] = "Nitrogen_Depletion_1h"
  names_Nitrogen_condition[3] = "Nitrogen_Depletion_2h"
  names_Nitrogen_condition[4] = "Nitrogen_Depletion_4h"
  names_Nitrogen_condition[5] = "Nitrogen_Depletion_8h"
  names_Nitrogen_condition[6] = "Nitrogen_Depletion_12h"
  names_Nitrogen_condition[7] = "Nitrogen_Depletion_1d"
}

{
  names_diamide_condition = vector(mode="character", length=8)
  names_diamide_condition[1] = "1.5mM_diamide_5min"
  names_diamide_condition[2] = "1.5mM_diamide_10min"
  names_diamide_condition[3] = "1.5mM_diamide_20min"
  names_diamide_condition[4] = "1.5mM_diamide_30min"
  names_diamide_condition[5] = "1.5mM_diamide_40min"
  names_diamide_condition[6] = "1.5mM_diamide_50min"
  names_diamide_condition[7] = "1.5mM_diamide_60min"
  names_diamide_condition[8] = "1.5mM_diamide_90min"
}




{
  names_dithiothrietol_condition = vector(mode="character", length=8)
  names_dithiothrietol_condition[1] = "2.5mM_DTT_005min_dtt-1"
  names_dithiothrietol_condition[2] = "2.5mM_DTT_015min_dtt-1"
  names_dithiothrietol_condition[3] = "2.5mM_DTT_030min_dtt-1"
  names_dithiothrietol_condition[4] = "2.5mM_DTT_045min_dtt-1"
  names_dithiothrietol_condition[5] = "2.5mM_DTT_060min_dtt-1"
  names_dithiothrietol_condition[6] = "2.5mM_DTT_090min_dtt-1"
  names_dithiothrietol_condition[7] = "2.5mM_DTT_120min_dtt-1"
  names_dithiothrietol_condition[8] = "2.5mM_DTT_180min_dtt-1"
}

{
  names_meanadione_condition = vector(mode="character", length=8)
  names_meanadione_condition[1] = "1mM_Menadione_10min_redo"
  names_meanadione_condition[2] = "1mM_Menadione_20min_redo"
  names_meanadione_condition[3] = "1mM_Menadione_30min_redo"
  names_meanadione_condition[4] = "1mM_Menadione_40min_redo"
  names_meanadione_condition[5] = "1mM_Menadione_50min_redo"
  names_meanadione_condition[6] = "1mM_Menadione_80min_redo"
  names_meanadione_condition[7] = "1mM_Menadione_120min_redo"
  names_meanadione_condition[8] = "1mM_Menadione_160min_redo"
}

{
  names_H2O2_condition = vector(mode="character", length=10)
  names_H2O2_condition[1] = "0.32mM_H2O2_10min_redo"
  names_H2O2_condition[2] = "0.32mM_H2O2_20min_redo"
  names_H2O2_condition[3] = "0.32mM_H2O2_30min_redo"
  names_H2O2_condition[4] = "0.32mM_H2O2_40min_rescan"
  names_H2O2_condition[5] = "0.32mM_H2O2_50min_redo"
  names_H2O2_condition[6] = "0.32mM_H2O2_60min_redo"
  names_H2O2_condition[7] = "0.32mM_H2O2_80min_redo"
  names_H2O2_condition[8] = "0.32mM_H2O2_100min_redo"
  names_H2O2_condition[9] = "0.32mM_H2O2_120min_redo"
  names_H2O2_condition[10] = "0.32mM_H2O2_160min_redo"
}

```


```{r echo=FALSE}

summary_coefficient_rank_average = data.frame(matrix(ncol = 3, nrow = 0))


temp_coefficient_rank = summary_coefficient_rank[summary_coefficient_rank$condition %in% names_heatshock_25to37_condition,]
for (motif in names_motifs_valid) {
  temp_coefficient_rank_2 = temp_coefficient_rank[temp_coefficient_rank$motif == motif,]
  
  summary_coefficient_rank_average = rbind(summary_coefficient_rank_average, data.frame(
    motif = motif,
    rank = sum(temp_coefficient_rank_2$model.L2*temp_coefficient_rank_2$rank)/
      sum(temp_coefficient_rank_2$model.L2),
    group = "heat shock"
  ))
}




temp_coefficient_rank = summary_coefficient_rank[summary_coefficient_rank$condition %in% names_heatshock_37to25_condition,]
for (motif in names_motifs_valid) {
  temp_coefficient_rank_2 = temp_coefficient_rank[temp_coefficient_rank$motif == motif,]
  
  summary_coefficient_rank_average = rbind(summary_coefficient_rank_average, data.frame(
    motif = motif,
    rank = sum(temp_coefficient_rank_2$model.L2*temp_coefficient_rank_2$rank)/
      sum(temp_coefficient_rank_2$model.L2),
    group = "hypothermia"
  ))
}


temp_coefficient_rank = summary_coefficient_rank[summary_coefficient_rank$condition %in% names_heatshock_sorbitol_condition,]
for (motif in names_motifs_valid) {
  temp_coefficient_rank_2 = temp_coefficient_rank[temp_coefficient_rank$motif == motif,]
  
  summary_coefficient_rank_average = rbind(summary_coefficient_rank_average, data.frame(
    motif = motif,
    rank = sum(temp_coefficient_rank_2$model.L2*temp_coefficient_rank_2$rank)/
      sum(temp_coefficient_rank_2$model.L2),
    group = "heat shock sorbitol"
  ))
}


temp_coefficient_rank = summary_coefficient_rank[summary_coefficient_rank$condition %in% names_sorbitol_condition,]
for (motif in names_motifs_valid) {
  temp_coefficient_rank_2 = temp_coefficient_rank[temp_coefficient_rank$motif == motif,]
  
  summary_coefficient_rank_average = rbind(summary_coefficient_rank_average, data.frame(
    motif = motif,
    rank = sum(temp_coefficient_rank_2$model.L2*temp_coefficient_rank_2$rank)/
      sum(temp_coefficient_rank_2$model.L2),
    group = "sorbitol"
  ))
}



temp_coefficient_rank = summary_coefficient_rank[summary_coefficient_rank$condition %in% names_diamide_condition,]
for (motif in names_motifs_valid) {
  temp_coefficient_rank_2 = temp_coefficient_rank[temp_coefficient_rank$motif == motif,]
  
  summary_coefficient_rank_average = rbind(summary_coefficient_rank_average, data.frame(
    motif = motif,
    rank = sum(temp_coefficient_rank_2$model.L2*temp_coefficient_rank_2$rank)/
      sum(temp_coefficient_rank_2$model.L2),
    group = "diamide"
  ))
}



temp_coefficient_rank = summary_coefficient_rank[summary_coefficient_rank$condition %in% names_Nitrogen_condition,]
for (motif in names_motifs_valid) {
  temp_coefficient_rank_2 = temp_coefficient_rank[temp_coefficient_rank$motif == motif,]
  
  summary_coefficient_rank_average = rbind(summary_coefficient_rank_average, data.frame(
    motif = motif,
    rank = sum(temp_coefficient_rank_2$model.L2*temp_coefficient_rank_2$rank)/
      sum(temp_coefficient_rank_2$model.L2),
    group = "nitrogen"
  ))
}


  
temp_coefficient_rank = summary_coefficient_rank[summary_coefficient_rank$condition %in% names_dithiothrietol_condition,]
for (motif in names_motifs_valid) {
  temp_coefficient_rank_2 = temp_coefficient_rank[temp_coefficient_rank$motif == motif,]
  
  summary_coefficient_rank_average = rbind(summary_coefficient_rank_average, data.frame(
    motif = motif,
    rank = sum(temp_coefficient_rank_2$model.L2*temp_coefficient_rank_2$rank)/
      sum(temp_coefficient_rank_2$model.L2),
    group = "dithiothrietol"
  ))
}


temp_coefficient_rank = summary_coefficient_rank[summary_coefficient_rank$condition %in% names_meanadione_condition,]
for (motif in names_motifs_valid) {
  temp_coefficient_rank_2 = temp_coefficient_rank[temp_coefficient_rank$motif == motif,]
  
  summary_coefficient_rank_average = rbind(summary_coefficient_rank_average, data.frame(
    motif = motif,
    rank = sum(temp_coefficient_rank_2$model.L2*temp_coefficient_rank_2$rank)/
      sum(temp_coefficient_rank_2$model.L2),
    group = "meanadione"
  ))
}

temp_coefficient_rank = summary_coefficient_rank[summary_coefficient_rank$condition %in% names_H2O2_condition,]
for (motif in names_motifs_valid) {
  temp_coefficient_rank_2 = temp_coefficient_rank[temp_coefficient_rank$motif == motif,]
  
  summary_coefficient_rank_average = rbind(summary_coefficient_rank_average, data.frame(
    motif = motif,
    rank = sum(temp_coefficient_rank_2$model.L2*temp_coefficient_rank_2$rank)/
      sum(temp_coefficient_rank_2$model.L2),
    group = "H2O2"
  ))
}


```


##### order the motif
```{r echo=FALSE}
order_motifs = summary_motif_frequency[order(-summary_motif_frequency$transcriptAmount),]$motif
summary_coefficient_rank_average$motif = factor(summary_coefficient_rank_average$motif, levels = order_motifs)

summary_coefficient_rank_average = merge(x = summary_coefficient_rank_average, y = summary_motif_frequency, by = "motif")

```


##### plot graph
```{r echo=FALSE, fig.height=8, fig.width=14}
ggplot(data=summary_coefficient_rank_average[summary_coefficient_rank_average$motif %in% head(order_motifs,31),], aes(x=motif)) +
  geom_bar(stat="identity", aes(y=1/rank,fill = group)) +
  geom_line(aes(y = transcriptAmount/2, colour = "log10 transcript frequency"),group=1, size=1.5) +
  geom_text(aes(y = transcriptAmount/2, label = round(transcriptAmount, 1)), vjust = -1.0, color = "black", size = 4) +
  theme(axis.text.x = element_text(size = 16, angle = 70, hjust = 1),
        axis.text.y = element_text(size = 16),
        legend.text = element_text(size = 16),
        axis.title = element_text(size = 16)) + 
  scale_x_discrete(name ="Motif") + 
  scale_colour_manual(name="", values=c("log10 transcript frequency"="gray20"))+
  scale_y_continuous("1 / weighted rank", sec.axis = sec_axis(~.*2, name = "Log10 transcript frequency"), limits = c(0,2)) +
  ggtitle("average rank motifs from different group of environmental stresses")+
  scale_fill_manual(values=c('#333333','#ACE5EE','#03C03C','#E69F00','#3C69E7','#999999','#E30022','#9933CC','#999966'))

```

```{r echo=FALSE, fig.height=8, fig.width=14}
ggplot(data=summary_coefficient_rank_average[summary_coefficient_rank_average$motif %in% tail(order_motifs,31),], aes(x=motif)) +
  geom_bar(stat="identity", aes(y=1/rank,fill = group)) +
  geom_line(aes(y = transcriptAmount/2, colour = "log10 transcript frequency"),group=1, size=1.5) +
  geom_text(aes(y = transcriptAmount/2, label = round(transcriptAmount, 1)), vjust = -1.0, color = "black", size = 4) +
  theme(axis.text.x = element_text(size = 16, angle = 70, hjust = 1),
        axis.text.y = element_text(size = 16),
        legend.text = element_text(size = 16),
        axis.title = element_text(size = 16)) + 
  scale_x_discrete(name ="Motif") + 
  scale_colour_manual(name="", values=c("log10 transcript frequency"="gray20"))+
  scale_y_continuous("1 / weighted rank", sec.axis = sec_axis(~.*2, name = "Log10 transcript frequency"), limits = c(0,2)) +
  ggtitle("average rank motifs from different group of environmental stresses")+
  scale_fill_manual(values=c('#333333','#ACE5EE','#03C03C','#E69F00','#3C69E7','#999999','#E30022','#9933CC','#999966'))
```


### <a id="bias"> investigating bias and expression mean of each model </a>
The bias of the model can be considered as the expected gene expression level change of a gene with no cis-regulatory element we considered in the model. In other words, the bias indicates the expression level change that cannot be explained by the model and could indicate the estimated proportion of the efforts on regulation these cis-regulatory elements make. 

```{r echo=FALSE}
summary_bias = data.frame(matrix(ncol = 3, nrow = 0))

biases = models_all_minLambda$a0

expressionLevel_matrix = as.matrix(expressionLevel_Gasch[4:length(expressionLevel_Gasch)])
storage.mode(expressionLevel_matrix) <- "double"
expressionLevel_column_mean = colMeans(expressionLevel_matrix,na.rm = TRUE)

i = 1
while(i <= length(names_environmental_stress)){
  summary_bias = rbind(summary_bias, data.frame(
    group = "bias",
    value = biases[i],
    condition = names_environmental_stress[i]
  ))
  summary_bias = rbind(summary_bias, data.frame(
    group = "expression mean",
    value = expressionLevel_column_mean[i],
    condition = names_environmental_stress[i]
  ))
  i = i+1
}
```


```{r echo=FALSE, fig.height=8, fig.width=14}
{
  names_heatshock_25to37_condition = vector(mode="character", length=8)
  names_heatshock_25to37_condition[1] = "hs_05min_hs-1"
  names_heatshock_25to37_condition[2] = "hs_10min_hs-1"
  names_heatshock_25to37_condition[3] = "hs_15min_hs-1"
  names_heatshock_25to37_condition[4] = "hs_20min_hs-1"
  names_heatshock_25to37_condition[5] = "hs_30min_hs-1"
  names_heatshock_25to37_condition[6] = "hs_40min_hs-1"
  names_heatshock_25to37_condition[7] = "hs_60min_hs-1"
  names_heatshock_25to37_condition[8] = "hs_80min_hs-1"
}

ggplot(data=summary_bias[summary_bias$condition %in% names_heatshock_25to37_condition,], aes(y = value, x=condition, group = group)) +
  geom_hline(yintercept=0, color = "gray", size = 2) +
  geom_line(aes(colour=group), size = 2) +
  theme(axis.text.x = element_text(size = 16, angle = 70, hjust = 1),
        axis.text.y = element_text(size = 16),
        legend.text = element_text(size = 16),
        axis.title = element_text(size = 16)) + 
  scale_x_discrete(name ="condition") + 
  ylim(-0.3,0.3)+
  ggtitle("bias of the models of heat shock")

```





```{r echo=FALSE, fig.height=8, fig.width=14}
{
  names_heatshock_37to25_condition = vector(mode="character", length=5)
  names_heatshock_37to25_condition[1] = "hs_37to25_15min"
  names_heatshock_37to25_condition[2] = "hs_37to25_30min"
  names_heatshock_37to25_condition[3] = "hs_37to25_45min"
  names_heatshock_37to25_condition[4] = "hs_37to25_60min"
  names_heatshock_37to25_condition[5] = "hs_37to25_90min"
}

ggplot(data=summary_bias[summary_bias$condition %in% names_heatshock_37to25_condition,], aes(y = value, x=condition, group = group)) +
  geom_hline(yintercept=0, color = "gray", size = 2) +
  geom_line(aes(colour=group), size = 2) +
  theme(axis.text.x = element_text(size = 16, angle = 70, hjust = 1),
        axis.text.y = element_text(size = 16),
        legend.text = element_text(size = 16),
        axis.title = element_text(size = 16)) + 
  scale_x_discrete(name ="condition") + 
  ylim(-0.3,0.3)+
  ggtitle("bias of the models of hypothermia")

```


```{r echo=FALSE, fig.height=11, fig.width=14}
{
  names_heatshock_sorbitol_condition = vector(mode="character", length=3)
  names_heatshock_sorbitol_condition[1] = "29C(1M_sorbitol)~33C(1M_sorbitol)_05min"
  names_heatshock_sorbitol_condition[2] = "29C(1M_sorbitol)~33C(1M_sorbitol)_15min"
  names_heatshock_sorbitol_condition[3] = "29C(1M_sorbitol)~33C(1M_sorbitol)_30min"
}

ggplot(data=summary_bias[summary_bias$condition %in% names_heatshock_sorbitol_condition,], aes(y = value, x=condition, group = group)) +
  geom_hline(yintercept=0, color = "gray", size = 2) +
  geom_line(aes(colour=group), size = 2) +
  theme(axis.text.x = element_text(size = 16, angle = 70, hjust = 1),
        axis.text.y = element_text(size = 16),
        legend.text = element_text(size = 16),
        axis.title = element_text(size = 16)) + 
  scale_x_discrete(name ="condition") + 
  ylim(-0.3,0.3)+
  ggtitle("bias of the models of heat shock with sorbitol")

```

```{r echo=FALSE, fig.height=8, fig.width=14}
{
  names_sorbitol_condition = vector(mode="character", length=7)
  names_sorbitol_condition[1] = "1M_sorbitol_05min"
  names_sorbitol_condition[2] = "1M_sorbitol_15min"
  names_sorbitol_condition[3] = "1M_sorbitol_30min"
  names_sorbitol_condition[4] = "1M_sorbitol_45min"
  names_sorbitol_condition[5] = "1M_sorbitol_60min"
  names_sorbitol_condition[6] = "1M_sorbitol_90min"
  names_sorbitol_condition[7] = "1M_sorbitol_120min"
}

ggplot(data=summary_bias[summary_bias$condition %in% names_sorbitol_condition,], aes(y = value, x=condition, group = group)) +
  geom_line(aes(colour=group), size = 2) +
  geom_hline(yintercept=0, color = "gray", size = 2) +
  theme(axis.text.x = element_text(size = 16, angle = 70, hjust = 1),
        axis.text.y = element_text(size = 16),
        legend.text = element_text(size = 16),
        axis.title = element_text(size = 16)) + 
  scale_x_discrete(name ="condition") + 
  ylim(-0.3,0.3)+
  ggtitle("bias of the models of sorbitol")

```




```{r echo=FALSE, fig.height=8, fig.width=14}
{
  names_Nitrogen_condition = vector(mode="character", length=7)
  names_Nitrogen_condition[1] = "Nitrogen_Depletion_30min"
  names_Nitrogen_condition[2] = "Nitrogen_Depletion_1h"
  names_Nitrogen_condition[3] = "Nitrogen_Depletion_2h"
  names_Nitrogen_condition[4] = "Nitrogen_Depletion_4h"
  names_Nitrogen_condition[5] = "Nitrogen_Depletion_8h"
  names_Nitrogen_condition[6] = "Nitrogen_Depletion_12h"
  names_Nitrogen_condition[7] = "Nitrogen_Depletion_1d"
}




ggplot(data=summary_bias[summary_bias$condition %in% names_Nitrogen_condition,], aes(y = value, x=condition, group = group)) +
  geom_hline(yintercept=0, color = "gray", size = 2) +
  geom_line(aes(colour=group), size = 2) +
  theme(axis.text.x = element_text(size = 16, angle = 70, hjust = 1),
        axis.text.y = element_text(size = 16),
        legend.text = element_text(size = 16),
        axis.title = element_text(size = 16)) + 
  scale_x_discrete(name ="condition") + 
  ylim(-0.3,0.3)+
  ggtitle("bias of the models of Nitrogen stress")
```


```{r echo=FALSE, fig.height=8, fig.width=14}
{
  names_diamide_condition = vector(mode="character", length=8)
  names_diamide_condition[1] = "1.5mM_diamide_5min"
  names_diamide_condition[2] = "1.5mM_diamide_10min"
  names_diamide_condition[3] = "1.5mM_diamide_20min"
  names_diamide_condition[4] = "1.5mM_diamide_30min"
  names_diamide_condition[5] = "1.5mM_diamide_40min"
  names_diamide_condition[6] = "1.5mM_diamide_50min"
  names_diamide_condition[7] = "1.5mM_diamide_60min"
  names_diamide_condition[8] = "1.5mM_diamide_90min"
}

ggplot(data=summary_bias[summary_bias$condition %in% names_diamide_condition,], aes(y = value, x=condition, group = group)) +
  geom_line(aes(colour=group), size = 2) +
  geom_hline(yintercept=0, color = "gray", size = 2) +
  theme(axis.text.x = element_text(size = 16, angle = 70, hjust = 1),
        axis.text.y = element_text(size = 16),
        legend.text = element_text(size = 16),
        axis.title = element_text(size = 16)) + 
  scale_x_discrete(name ="condition") + 
  ylim(-0.3,0.3)+
  ggtitle("bias of the models of diamide stress")

```






### <a id="weightedL2"> Select motif with weighted square L2 </a>
In order to focus only on the significant motifs, we calculate the `variance predict by motif` for each motif over all the conditions. In this case, we consider no only the regression coefficient but also the transcript frequence: 
$$
\text{c: condition}\\
\text{g: gene}\\
\text{G: number of genes}\\[2ex]
e_{mg}=
\begin{cases}
1,\quad \text{at least one motif } m \text{ is on gene }g\\
0, \quad otherwise
\end{cases}
\\[6ex]

\begin{align}
\text{Variance pridict by motif} &=   \sum_c\sum_g \beta _{mc}^2 \times e_{mg}\\
&= \beta^2_{mc} \times transcript\ frequency \\
&= L_2^2\times transcript\ frequency
\end{align}

$$



##### group conditions
```{r}

{
  names_heatshock_29to33_condition = vector(mode="character", length=3)
  names_heatshock_29to33_condition[1] = "hs_29to33_05min"
  names_heatshock_29to33_condition[2] = "hs_29to33_15min"
  names_heatshock_29to33_condition[3] = "hs_29to33_30min"
}


{
  names_heatshock_sorbitol_condition = vector(mode="character", length=3)
  names_heatshock_sorbitol_condition[1] = "29C(1M_sorbitol)~33C(1M_sorbitol)_05min"
  names_heatshock_sorbitol_condition[2] = "29C(1M_sorbitol)~33C(1M_sorbitol)_15min"
  names_heatshock_sorbitol_condition[3] = "29C(1M_sorbitol)~33C(1M_sorbitol)_30min"
}


{
  names_heatshock_25to37_condition = vector(mode="character", length=8)
  names_heatshock_25to37_condition[1] = "hs_05min_hs-1"
  names_heatshock_25to37_condition[2] = "hs_10min_hs-1"
  names_heatshock_25to37_condition[3] = "hs_15min_hs-1"
  names_heatshock_25to37_condition[4] = "hs_20min_hs-1"
  names_heatshock_25to37_condition[5] = "hs_30min_hs-1"
  names_heatshock_25to37_condition[6] = "hs_40min_hs-1"
  names_heatshock_25to37_condition[7] = "hs_60min_hs-1"
  names_heatshock_25to37_condition[8] = "hs_80min_hs-1"
}


{
  names_heatshock_37to25_condition = vector(mode="character", length=5)
  names_heatshock_37to25_condition[1] = "hs_37to25_15min"
  names_heatshock_37to25_condition[2] = "hs_37to25_30min"
  names_heatshock_37to25_condition[3] = "hs_37to25_45min"
  names_heatshock_37to25_condition[4] = "hs_37to25_60min"
  names_heatshock_37to25_condition[5] = "hs_37to25_90min"
}


{
  names_sorbitol_condition = vector(mode="character", length=7)
  names_sorbitol_condition[1] = "1M_sorbitol_05min"
  names_sorbitol_condition[2] = "1M_sorbitol_15min"
  names_sorbitol_condition[3] = "1M_sorbitol_30min"
  names_sorbitol_condition[4] = "1M_sorbitol_45min"
  names_sorbitol_condition[5] = "1M_sorbitol_60min"
  names_sorbitol_condition[6] = "1M_sorbitol_90min"
  names_sorbitol_condition[7] = "1M_sorbitol_120min"
}


{
  names_Nitrogen_condition = vector(mode="character", length=7)
  names_Nitrogen_condition[1] = "Nitrogen_Depletion_30min"
  names_Nitrogen_condition[2] = "Nitrogen_Depletion_1h"
  names_Nitrogen_condition[3] = "Nitrogen_Depletion_2h"
  names_Nitrogen_condition[4] = "Nitrogen_Depletion_4h"
  names_Nitrogen_condition[5] = "Nitrogen_Depletion_8h"
  names_Nitrogen_condition[6] = "Nitrogen_Depletion_12h"
  names_Nitrogen_condition[7] = "Nitrogen_Depletion_1d"
}

{
  names_diamide_condition = vector(mode="character", length=8)
  names_diamide_condition[1] = "1.5mM_diamide_5min"
  names_diamide_condition[2] = "1.5mM_diamide_10min"
  names_diamide_condition[3] = "1.5mM_diamide_20min"
  names_diamide_condition[4] = "1.5mM_diamide_30min"
  names_diamide_condition[5] = "1.5mM_diamide_40min"
  names_diamide_condition[6] = "1.5mM_diamide_50min"
  names_diamide_condition[7] = "1.5mM_diamide_60min"
  names_diamide_condition[8] = "1.5mM_diamide_90min"
}



{
  names_dithiothrietol_condition = vector(mode="character", length=7)
  names_dithiothrietol_condition[1] = "2.5mM_DTT_005min_dtt-1"
  names_dithiothrietol_condition[2] = "2.5mM_DTT_015min_dtt-1"
  names_dithiothrietol_condition[3] = "2.5mM_DTT_030min_dtt-1"
  names_dithiothrietol_condition[4] = "2.5mM_DTT_045min_dtt-1"
  names_dithiothrietol_condition[5] = "2.5mM_DTT_060min_dtt-1"
  names_dithiothrietol_condition[6] = "2.5mM_DTT_090min_dtt-1"
  names_dithiothrietol_condition[7] = "2.5mM_DTT_120min_dtt-1"
  names_dithiothrietol_condition[8] = "2.5mM_DTT_180min_dtt-1"
}



{
  names_meanadione_condition = vector(mode="character", length=9)
  names_meanadione_condition[1] = "1mM_Menadione_10min_redo"
  names_meanadione_condition[2] = "1mM_Menadione_20min_redo"
  names_meanadione_condition[3] = "1mM_Menadione_30min_redo"
  names_meanadione_condition[4] = "1mM_Menadione_40min_redo"
  names_meanadione_condition[5] = "1mM_Menadione_50min_redo"
  names_meanadione_condition[6] = "1mM_Menadione_80min_redo"
  names_meanadione_condition[7] = "1mM_Menadione_105min_redo"
  names_meanadione_condition[8] = "1mM_Menadione_120min_redo"
  names_meanadione_condition[9] = "1mM_Menadione_160min_redo"
}

{
  names_H2O2_condition = vector(mode="character", length=9)
  names_H2O2_condition[1] = "0.32mM_H2O2_10min_redo"
  names_H2O2_condition[2] = "0.32mM_H2O2_20min_redo"
  names_H2O2_condition[3] = "0.32mM_H2O2_30min_redo"
  names_H2O2_condition[4] = "0.32mM_H2O2_40min_rescan"
  names_H2O2_condition[5] = "0.32mM_H2O2_50min_redo"
  names_H2O2_condition[6] = "0.32mM_H2O2_60min_redo"
  names_H2O2_condition[7] = "0.32mM_H2O2_80min_redo"
  names_H2O2_condition[8] = "0.32mM_H2O2_120min_redo"
  names_H2O2_condition[9] = "0.32mM_H2O2_160min_redo"
}


```


```{r echo=FALSE}
summary_motif_frequency_percentage = data.frame(matrix(ncol= 0, nrow = 1))
for(motif in names_motifs_valid){
  summary_motif_frequency_percentage[motif] = sum(fullTable_Gasch[[motif]] > 0)/nrow(fullTable_Gasch)
}
```



```{r echo=FALSE}

summary_coefficient = data.frame(matrix(ncol = 5, nrow = 0))

coefficients = extract_coefficient(models_all_minLambda,names_environmental_stress)

for(motif in names_motifs_valid){
  coefficients_motif_specific = coefficients[coefficients$rn == motif,]$value
  summary_coefficient = rbind(summary_coefficient, data.frame(
    motif = motif,
    condition = "all",
    L2 = sqrt(sum(coefficients_motif_specific^2)),
    meanL2 = sqrt(sum(coefficients_motif_specific^2) / length(names_environmental_stress)),
    varianceExplain = sum(coefficients_motif_specific^2*as.numeric(summary_motif_frequency_percentage[motif])) - sum((coefficients_motif_specific*as.numeric(summary_motif_frequency_percentage[motif]))^2)
  ))
}



coefficients = extract_coefficient(models_all_minLambda,names_heatshock_29to33_condition)

for(motif in names_motifs_valid){
  coefficients_motif_specific = coefficients[coefficients$rn == motif,]$value
  summary_coefficient = rbind(summary_coefficient, data.frame(
    motif = motif,
    condition = "heat shock from 29C to 33C",
    L2 = sqrt(sum(coefficients_motif_specific^2)),
    meanL2 = sqrt(sum(coefficients_motif_specific^2) / length(names_heatshock_29to33_condition)),
    varianceExplain = sum(coefficients_motif_specific^2*as.numeric(summary_motif_frequency_percentage[motif])) - sum((coefficients_motif_specific*as.numeric(summary_motif_frequency_percentage[motif]))^2)
  ))
}



coefficients = extract_coefficient(models_all_minLambda,names_heatshock_sorbitol_condition)

for(motif in names_motifs_valid){
  coefficients_motif_specific = coefficients[coefficients$rn == motif,]$value
  summary_coefficient = rbind(summary_coefficient, data.frame(
    motif = motif,
    condition = "heat shock from 29C to 33C with sorbitol",
    L2 = sqrt(sum(coefficients_motif_specific^2)),
    meanL2 = sqrt(sum(coefficients_motif_specific^2) / length(names_heatshock_sorbitol_condition)),
    varianceExplain = sum(coefficients_motif_specific^2*as.numeric(summary_motif_frequency_percentage[motif])) - sum((coefficients_motif_specific*as.numeric(summary_motif_frequency_percentage[motif]))^2)
  ))
}



coefficients = extract_coefficient(models_all_minLambda,names_heatshock_25to37_condition)

for(motif in names_motifs_valid){
  coefficients_motif_specific = coefficients[coefficients$rn == motif,]$value
  summary_coefficient = rbind(summary_coefficient, data.frame(
    motif = motif,
    condition = "heat shock from 25C to 37C",
    L2 = sqrt(sum(coefficients_motif_specific^2)),
    meanL2 = sqrt(sum(coefficients_motif_specific^2) / length(names_heatshock_25to37_condition)),
    varianceExplain = sum(coefficients_motif_specific^2*as.numeric(summary_motif_frequency_percentage[motif])) - sum((coefficients_motif_specific*as.numeric(summary_motif_frequency_percentage[motif]))^2)
  ))
}




coefficients = extract_coefficient(models_all_minLambda,names_heatshock_37to25_condition)

for(motif in names_motifs_valid){
  coefficients_motif_specific = coefficients[coefficients$rn == motif,]$value
  summary_coefficient = rbind(summary_coefficient, data.frame(
    motif = motif,
    condition = "hypothermia from 37C to 25C",
    L2 = sqrt(sum(coefficients_motif_specific^2)),
    meanL2 = sqrt(sum(coefficients_motif_specific^2) / length(names_heatshock_37to25_condition)),
    varianceExplain = sum(coefficients_motif_specific^2*as.numeric(summary_motif_frequency_percentage[motif])) - sum((coefficients_motif_specific*as.numeric(summary_motif_frequency_percentage[motif]))^2)
  ))
}



coefficients = extract_coefficient(models_all_minLambda,names_sorbitol_condition)

for(motif in names_motifs_valid){
  coefficients_motif_specific = coefficients[coefficients$rn == motif,]$value
  summary_coefficient = rbind(summary_coefficient, data.frame(
    motif = motif,
    condition = "sorbitol",
    L2 = sqrt(sum(coefficients_motif_specific^2)),
    meanL2 = sqrt(sum(coefficients_motif_specific^2) / length(names_sorbitol_condition)),
    varianceExplain = sum(coefficients_motif_specific^2*as.numeric(summary_motif_frequency_percentage[motif])) - sum((coefficients_motif_specific*as.numeric(summary_motif_frequency_percentage[motif]))^2)
  ))
}




coefficients = extract_coefficient(models_all_minLambda,names_Nitrogen_condition)

for(motif in names_motifs_valid){
  coefficients_motif_specific = coefficients[coefficients$rn == motif,]$value
  summary_coefficient = rbind(summary_coefficient, data.frame(
    motif = motif,
    condition = "nitrogen",
    L2 = sqrt(sum(coefficients_motif_specific^2)),
    meanL2 = sqrt(sum(coefficients_motif_specific^2) / length(names_Nitrogen_condition)),
    varianceExplain = sum(coefficients_motif_specific^2*as.numeric(summary_motif_frequency_percentage[motif])) - sum((coefficients_motif_specific*as.numeric(summary_motif_frequency_percentage[motif]))^2)
  ))
}




coefficients = extract_coefficient(models_all_minLambda,names_diamide_condition)

for(motif in names_motifs_valid){
  coefficients_motif_specific = coefficients[coefficients$rn == motif,]$value
  summary_coefficient = rbind(summary_coefficient, data.frame(
    motif = motif,
    condition = "diamide",
    L2 = sqrt(sum(coefficients_motif_specific^2)),
    meanL2 = sqrt(sum(coefficients_motif_specific^2) / length(names_diamide_condition)),
    varianceExplain = sum(coefficients_motif_specific^2*as.numeric(summary_motif_frequency_percentage[motif])) - sum((coefficients_motif_specific*as.numeric(summary_motif_frequency_percentage[motif]))^2)
  ))
}





coefficients = extract_coefficient(models_all_minLambda,names_dithiothrietol_condition)

for(motif in names_motifs_valid){
  coefficients_motif_specific = coefficients[coefficients$rn == motif,]$value
  summary_coefficient = rbind(summary_coefficient, data.frame(
    motif = motif,
    condition = "dithiothrietol",
    L2 = sqrt(sum(coefficients_motif_specific^2)),
    meanL2 = sqrt(sum(coefficients_motif_specific^2) / length(names_dithiothrietol_condition)),
    varianceExplain = sum(coefficients_motif_specific^2*as.numeric(summary_motif_frequency_percentage[motif])) - sum((coefficients_motif_specific*as.numeric(summary_motif_frequency_percentage[motif]))^2)
  ))
}


coefficients = extract_coefficient(models_all_minLambda,names_meanadione_condition)

for(motif in names_motifs_valid){
  coefficients_motif_specific = coefficients[coefficients$rn == motif,]$value
  summary_coefficient = rbind(summary_coefficient, data.frame(
    motif = motif,
    condition = "meanadione",
    L2 = sqrt(sum(coefficients_motif_specific^2)),
    meanL2 = sqrt(sum(coefficients_motif_specific^2) / length(names_meanadione_condition)),
    varianceExplain = sum(coefficients_motif_specific^2*as.numeric(summary_motif_frequency_percentage[motif])) - sum((coefficients_motif_specific*as.numeric(summary_motif_frequency_percentage[motif]))^2)
  ))
}



coefficients = extract_coefficient(models_all_minLambda,names_H2O2_condition)

for(motif in names_motifs_valid){
  coefficients_motif_specific = coefficients[coefficients$rn == motif,]$value
  summary_coefficient = rbind(summary_coefficient, data.frame(
    motif = motif,
    condition = "H2O2",
    L2 = sqrt(sum(coefficients_motif_specific^2)),
    meanL2 = sqrt(sum(coefficients_motif_specific^2) / length(names_H2O2_condition)),
    varianceExplain = sum(coefficients_motif_specific^2*as.numeric(summary_motif_frequency_percentage[motif])) - sum((coefficients_motif_specific*as.numeric(summary_motif_frequency_percentage[motif]))^2)
  ))
}


```

```{r echo=FALSE}
summary_motif_frequency = data.frame(matrix(ncol= 2, nrow = 0))

for(motif in names_motifs_valid){
  summary_motif_frequency = rbind(summary_motif_frequency, data.frame(
    motif = motif,
    transcriptAmount = sum(fullTable_Gasch[[motif]] > 0)
  ))
}

```


```{r echo = FALSE}
summary_coefficient = merge(x = summary_coefficient, y = summary_motif_frequency, by = "motif")
```

```{r echo=FALSE}
summary_coefficient$weightedsquareL2 = summary_coefficient$transcriptAmount*summary_coefficient$L2*summary_coefficient$L2
summary_coefficient$weightedsquaremeanL2 = summary_coefficient$transcriptAmount*summary_coefficient$meanL2*summary_coefficient$meanL2
```



##### manage plotting order
```{r echo=FALSE}
temp_summary_coefficient = summary_coefficient[summary_coefficient$condition == "all",]
order_motifs = temp_summary_coefficient[order(-temp_summary_coefficient$weightedsquaremeanL2),]$motif
summary_coefficient$motif = factor(summary_coefficient$motif, levels = order_motifs)
```

##### plotting the weighted square L2 of each motif
```{r echo=FALSE, fig.height=8, fig.width=14}
temp_summary_coefficient = summary_coefficient[summary_coefficient$condition == "all",]

ggplot(data=temp_summary_coefficient[temp_summary_coefficient$motif %in% head(order_motifs,31),], 
       aes(x=motif, y=weightedsquaremeanL2, fill=condition)) +
  geom_bar(stat="identity", position=position_dodge()) +
  theme(axis.text.x = element_text(size = 16, angle = 70, hjust = 1),
        axis.text.y = element_text(size = 16),
        legend.text = element_text(size = 16),
        axis.title = element_text(size = 16),
        title = element_text(size = 14)) + 
  scale_x_discrete(name ="Motif") + 
  ggtitle("weighted square mean L2 of the non-zero regression coefficient under all condition")
```



```{r echo=FALSE, fig.height=8, fig.width=14}
temp_summary_coefficient = summary_coefficient[summary_coefficient$condition == "all",]

ggplot(data=temp_summary_coefficient[temp_summary_coefficient$motif %in% head(order_motifs,31),], 
       aes(x=motif, y=weightedsquareL2, fill=condition)) +
  geom_bar(stat="identity", position=position_dodge()) +
  theme(axis.text.x = element_text(size = 16, angle = 70, hjust = 1),
        axis.text.y = element_text(size = 16),
        legend.text = element_text(size = 16),
        axis.title = element_text(size = 16),
        title = element_text(size = 14)) + 
  scale_x_discrete(name ="Motif") + 
  ggtitle("weighted square L2 of the non-zero regression coefficient under all condition")
```


##### select motifs
We selected the 6 motifs with the highest weighted L2 value. They should be reliable significant motifs under the environmental conditions we considered.

```{r}
names_motifs_significant = head(order_motifs,6)
```




### <a id="meanL2"> Compare mean of L2 of significant motifs across different group of environmental stress </a>
First, We align the models by only considering the model with time slices within 5min to 60mins. For each group of environmental stress, we calculate the mean of L2 to allow cross comparison across environmental stresses group with different number of models:

$$
mean\ L_2 = \sqrt{{\sum_c^N \beta _{mc}^2}\over \text{N}}
$$



```{r echo=FALSE, fig.height=8, fig.width=14}
temp_summary_coefficient = summary_coefficient[summary_coefficient$motif %in% names_motifs_significant[1:3] & summary_coefficient$condition != "all",]

ggplot(data=temp_summary_coefficient, aes(x=motif, y=meanL2, fill=condition)) +
  geom_bar(stat="identity", position=position_dodge()) +
  theme(axis.text.x = element_text(size = 16),
        axis.text.y = element_text(size = 16),
        legend.text = element_text(size = 16),
        axis.title = element_text(size = 16)) + 
  scale_x_discrete(name ="Motif") +
  ylim(0,2) +
  ggtitle("compare mean L2 of motif from different environmental stresses group")
```



```{r echo=FALSE, fig.height=8, fig.width=14}
temp_summary_coefficient = summary_coefficient[summary_coefficient$motif %in% names_motifs_significant[4:6] & summary_coefficient$condition != "all",]

ggplot(data=temp_summary_coefficient, aes(x=motif, y=meanL2, fill=condition)) +
  geom_bar(stat="identity", position=position_dodge()) +
  theme(axis.text.x = element_text(size = 16),
        axis.text.y = element_text(size = 16),
        legend.text = element_text(size = 16),
        axis.title = element_text(size = 16)) + 
  scale_x_discrete(name ="Motif") + 
  ylim(0,2) +
  ggtitle("compare mean L2 of motif from different environmental stresses group")
```


### <a id="proportionVar"> compare proportion of variance explain by each motif </a>
However, the overall regulatory effect in a cell also depends on how strong the environmental stress is. For example, the change in expression level under heat shock from 25 degree to 37 degree should be much larger than that under heat shock from 29 degree to 33 degree. Therefore, we calculate the proportion of variance explain by motifs over the total variance to reduce such difference

We plot bar graphs to compare the value across each group of environmental stresses. The shape of the proportion variance graph is similar to that of the mean of L2. There is still a large difference between the value of heat shock from 29C to 33C and heat shock from 25C to 37C.

$$
\begin{align}
\text{Variance pridict by motif} &=   \sum_c({1\over G}\times \sum_g \beta _{mc}^2 \times e_{mg}  - ({1\over G}\times\sum_g \beta _{mc} \times e_{mg})^2)\\
&= \sum_c (\beta^2_{mc}\ \times\ percentage\ of\ transcript\ frequence- (\beta_{mc} \times percentage\ of\ transcript\ frequence)^2) \\
\end{align}
$$

$$
\text{proportion of variance explain by motif } = {\text{variance explain by motifs}\over \text{total variance}}
$$

##### compute the total variance of expression level in each group
```{r echo=FALSE}

summary_variance = data.frame(matrix(ncol = 2, nrow = 0))

var = 0
for(condition in names_environmental_stress){
  var = var + var(fullTable_Gasch_impute_knn[condition])
}

summary_variance = rbind(summary_variance, data.frame(
  condition = "all",
  sumVariance = as.numeric(var)
))


var = 0
for(condition in names_heatshock_29to33_condition){
  var = var + var(fullTable_Gasch_impute_knn[condition])
}

summary_variance = rbind(summary_variance, data.frame(
  condition = "heat shock from 29C to 33C",
  sumVariance = as.numeric(var)
))


var = 0
for(condition in names_heatshock_sorbitol_condition){
  var = var + var(fullTable_Gasch_impute_knn[condition])
}

summary_variance = rbind(summary_variance, data.frame(
  condition = "heat shock from 29C to 33C with sorbitol",
  sumVariance = as.numeric(var)
))


var = 0
for(condition in names_heatshock_25to37_condition){
  var = var + var(fullTable_Gasch_impute_knn[condition])
}

summary_variance = rbind(summary_variance, data.frame(
  condition = "heat shock from 25C to 37C",
  sumVariance = as.numeric(var)
))


var = 0
for(condition in names_heatshock_37to25_condition){
  var = var + var(fullTable_Gasch_impute_knn[condition])
}

summary_variance = rbind(summary_variance, data.frame(
  condition = "hypothermia from 37C to 25C",
  sumVariance = as.numeric(var)
))


var = 0
for(condition in names_sorbitol_condition){
  var = var + var(fullTable_Gasch_impute_knn[condition])
}

summary_variance = rbind(summary_variance, data.frame(
  condition = "sorbitol",
  sumVariance = as.numeric(var)
))



var = 0
for(condition in names_Nitrogen_condition){
  var = var + var(fullTable_Gasch_impute_knn[condition])
}

summary_variance = rbind(summary_variance, data.frame(
  condition = "nitrogen",
  sumVariance = as.numeric(var)
))


var = 0
for(condition in names_diamide_condition){
  var = var + var(fullTable_Gasch_impute_knn[condition])
}

summary_variance = rbind(summary_variance, data.frame(
  condition = "diamide",
  sumVariance = as.numeric(var)
))



var = 0
for(condition in names_dithiothrietol_condition){
  var = var + var(fullTable_Gasch_impute_knn[condition])
}

summary_variance = rbind(summary_variance, data.frame(
  condition = "dithiothrietol",
  sumVariance = as.numeric(var)
))



var = 0
for(condition in names_meanadione_condition){
  var = var + var(fullTable_Gasch_impute_knn[condition])
}

summary_variance = rbind(summary_variance, data.frame(
  condition = "meanadione",
  sumVariance = as.numeric(var)
))


var = 0
for(condition in names_H2O2_condition){
  var = var + var(fullTable_Gasch_impute_knn[condition])
}

summary_variance = rbind(summary_variance, data.frame(
  condition = "H2O2",
  sumVariance = as.numeric(var)
))


```


```{r echo=FALSE}
summary_coefficient = merge(x = summary_coefficient, y = summary_variance, by = "condition", all = TRUE)
```


```{r echo=FALSE}
summary_coefficient$proportionVariance = summary_coefficient$varianceExplain/summary_coefficient$sumVariance
```


```{r echo=FALSE, fig.height=8, fig.width=14}
temp_summary_coefficient = summary_coefficient[summary_coefficient$motif %in% names_motifs_significant[1:3] & summary_coefficient$condition != "all",]

ggplot(data=temp_summary_coefficient, aes(x=motif, y=proportionVariance, fill=condition)) +
  geom_bar(stat="identity", position=position_dodge()) +
  theme(axis.text.x = element_text(size = 16),
        axis.text.y = element_text(size = 16),
        legend.text = element_text(size = 16),
        axis.title = element_text(size = 16)) + 
  scale_x_discrete(name ="Motif") + 
  ylim(0,0.03) +
  ggtitle("compare proportion variance of motif from different environmental stresses group")
```

```{r echo=FALSE, fig.height=8, fig.width=14}
temp_summary_coefficient = summary_coefficient[summary_coefficient$motif %in% names_motifs_significant[4:6] & summary_coefficient$condition != "all",]

ggplot(data=temp_summary_coefficient, aes(x=motif, y=proportionVariance, fill=condition)) +
  geom_bar(stat="identity", position=position_dodge()) +
  theme(axis.text.x = element_text(size = 16),
        axis.text.y = element_text(size = 16),
        legend.text = element_text(size = 16),
        axis.title = element_text(size = 16)) + 
  scale_x_discrete(name ="Motif") + 
  ylim(0,0.03) +
  ggtitle("compare proportion variance of motif from different environmental stresses group")
```









### <a id="linearSignificant"> significant test with linear model </a>
To evaluate the reliability of the regression coefficient of significant motifs, we fit a linear model with selected motifs and investigate their regression coefficient and standard error. According to previous figures, we apply the gene expression levels 15 or 20 mins after specific environmental stress when the regulatory effect of motifs is strong. We fitted 10 models in total and plot their coefficient and standard error. The error bars represent 1 standard error.


```{r}
 names_motifs_significant = c(
  'UGUAHMNUA',
  'ATATTC',
  'TGTATAWT',
  'TGTAAATA',
  'WUUGUAWUWU',
  'ATACGGTGT'
)
```

##### heat shock from 29C to 33C
```{r echo=FALSE, fig.height=6, fig.width=8}
formula_model = as.formula(paste("hs_29to33_15min ~",paste(names_motifs_significant, collapse = "+")))
model_hs_29to33 = lm(formula_model,fullTable_Gasch_impute_knn)

par(mfrow=c(2,2))
plot(model_hs_29to33)
```

##### heat shock from 29C to 33C with sorbitol
```{r echo=FALSE, fig.height=6, fig.width=8}
fullTable_Gasch_impute_knn$temp = fullTable_Gasch_impute_knn$`29C(1M_sorbitol)~33C(1M_sorbitol)_15min`
formula_model = as.formula(paste("temp ~",paste(names_motifs_significant, collapse = "+")))
model_hs_29to33_sorbitol = lm(formula_model,fullTable_Gasch_impute_knn)

par(mfrow=c(2,2))
plot(model_hs_29to33_sorbitol)
```

##### heat shock from 25C to 37C
```{r echo=FALSE, fig.height=6, fig.width=8}
fullTable_Gasch_impute_knn$temp = fullTable_Gasch_impute_knn$`hs_15min_hs-1`
formula_model = as.formula(paste("temp ~",paste(names_motifs_significant, collapse = "+")))
model_hs_25to37 = lm(formula_model,fullTable_Gasch_impute_knn)

par(mfrow=c(2,2))
plot(model_hs_25to37)
```

##### hypothermia from 37C to 25C
```{r echo=FALSE, fig.height=6, fig.width=8}
formula_model = as.formula(paste("hs_37to25_15min ~",paste(names_motifs_significant, collapse = "+")))
model_hs_37to25 = lm(formula_model,fullTable_Gasch_impute_knn)

par(mfrow=c(2,2))
plot(model_hs_37to25)
```


##### sorbitol
```{r echo=FALSE, fig.height=6, fig.width=8}
fullTable_Gasch_impute_knn$temp = fullTable_Gasch_impute_knn$`1M_sorbitol_30min`
formula_model = as.formula(paste("temp ~",paste(names_motifs_significant, collapse = "+")))
model_sorbitol = lm(formula_model,fullTable_Gasch_impute_knn)

par(mfrow=c(2,2))
plot(model_sorbitol)
```


##### nitrogen
```{r echo=FALSE, fig.height=6, fig.width=8}
formula_model = as.formula(paste("Nitrogen_Depletion_2d ~",paste(names_motifs_significant, collapse = "+")))
model_nitrogen = lm(formula_model,fullTable_Gasch_impute_knn)

par(mfrow=c(2,2))
plot(model_nitrogen)
```

##### dianide
```{r echo=FALSE, fig.height=6, fig.width=8}
fullTable_Gasch_impute_knn$temp = fullTable_Gasch_impute_knn$`1.5mM_diamide_30min`
formula_model = as.formula(paste("temp ~",paste(names_motifs_significant, collapse = "+")))
model_diamide = lm(formula_model,fullTable_Gasch_impute_knn)

par(mfrow=c(2,2))
plot(model_diamide)
```


##### dithiothriethol
```{r echo=FALSE, fig.height=6, fig.width=8}
fullTable_Gasch_impute_knn$temp = fullTable_Gasch_impute_knn$`2.5mM_DTT_045min_dtt-1`
formula_model = as.formula(paste("temp ~",paste(names_motifs_significant, collapse = "+")))
model_dithiothriethol = lm(formula_model,fullTable_Gasch_impute_knn)

par(mfrow=c(2,2))
plot(model_dithiothriethol)
```


##### menadione
```{r echo=FALSE, fig.height=6, fig.width=8}
fullTable_Gasch_impute_knn$temp = fullTable_Gasch_impute_knn$`1mM_Menadione_30min_redo`
formula_model = as.formula(paste("temp ~",paste(names_motifs_significant, collapse = "+")))
model_menadione = lm(formula_model,fullTable_Gasch_impute_knn)


par(mfrow=c(2,2))
plot(model_menadione)
```


##### H2O2
```{r echo=FALSE, fig.height=6, fig.width=8}
fullTable_Gasch_impute_knn$temp = fullTable_Gasch_impute_knn$`0.32mM_H2O2_30min_redo`
formula_model = as.formula(paste("temp ~",paste(names_motifs_significant, collapse = "+")))
model_H2O2 = lm(formula_model,fullTable_Gasch_impute_knn)

par(mfrow=c(2,2))
plot(model_H2O2)
```




```{r echo=FALSE}
summary_coefficient = rbind(
  extract_coefficient_lm(model_hs_29to33,"heat shock from 29C to 33C"),
  extract_coefficient_lm(model_hs_29to33_sorbitol,"heat shock from 29C to 33C with sorbitol"),
  extract_coefficient_lm(model_hs_25to37,"heat shock from 25C to 37C"),
  extract_coefficient_lm(model_hs_37to25,"hypothermia from 37C to 25C"),
  extract_coefficient_lm(model_sorbitol,"sorbitol"),
  extract_coefficient_lm(model_nitrogen,"nitrogen"),
  extract_coefficient_lm(model_diamide,"diamide"),
  extract_coefficient_lm(model_dithiothriethol,"dithiothriethol"),
  extract_coefficient_lm(model_menadione,"menadione"),
  extract_coefficient_lm(model_H2O2,"H2O2")
  )

```



##### <a id="significantCoefficient"> coefficients </a>
```{r echo=FALSE, fig.height=8, fig.width=14}
ggplot(summary_coefficient[summary_coefficient$motif %in% names_motifs_significant[1:3],], aes(x=motif, y=Estimate, group=condition, color=condition)) + 
  geom_hline(yintercept=0, linetype="dashed", size = 2, color = "gray50")+
  geom_point(position=position_dodge(0.6),size = 3)+
  geom_errorbar(aes(ymin=Estimate-2*Std..Error, ymax=Estimate+2*Std..Error),width=.8,size=1,position=position_dodge(0.6))+
  theme(axis.text.x = element_text(size = 16),
        axis.text.y = element_text(size = 16),
        legend.text = element_text(size = 16),
        axis.title = element_text(size = 16)) + 
  ylim(-1.1,1)+
  ggtitle("linear regression coefficient of significant motifs under each environmental stress")

```


```{r echo=FALSE, fig.height=8, fig.width=14}
ggplot(summary_coefficient[summary_coefficient$motif %in% names_motifs_significant[4:6],], aes(x=motif, y=Estimate, group=condition, color=condition)) + 
  geom_hline(yintercept=0, linetype="dashed", size = 2, color = "gray50")+
  geom_point(position=position_dodge(0.6),size = 3)+
  geom_errorbar(aes(ymin=Estimate-2*Std..Error, ymax=Estimate+2*Std..Error), width=.8, size=1,position=position_dodge(0.6))+
  theme(axis.text.x = element_text(size = 16),
        axis.text.y = element_text(size = 16),
        legend.text = element_text(size = 16),
        axis.title = element_text(size = 16)) + 
  ylim(-1,1.1)+
  ggtitle("linear regression coefficient of significant motifs under each environmental stress")

```







