---
title: "Abhi Project Report"
author: "Abhishek Jain"
date: "10/09/2019"
output: html_document
editor_options: 
  chunk_output_type: console
---


### Setup (not shown)

Includes libraries

```{R setup, warning=TRUE, include=FALSE}
require("knitr")
#set up working directory
  opts_knit$set(root.dir = "..")

if(!require(tidyverse)){
    install.packages("tidyverse")
    library(tidyverse)
}
if(!require(ggplot2)){
    install.packages("ggplot2")
    library(tidyverse)
}
if(!require(readxl)){
    install.packages("readxl")
    library(readxl)
}
#theme for ggplot  
  theme_coding <- function(...){
  theme_bw()+
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title = element_text(size = 12),
        panel.grid = element_blank(),
        plot.margin = unit(c(0.5, 0.5, 0.5, 0.5), units = , "cm"),
        plot.title = element_text(size = 20, vjust = 1, hjust = 0.5),
        legend.text = element_text(size = 12),
        legend.title = element_blank(),
        legend.position = c(0.9, 0.9))
}

#defining fun color for scheming
colour <- c(teal = "#237E8A", green = "#1A8A21", red = "#A62621")

```

---

### Loading Base Data

Following code loads the necessary data. Note that reference data loaded here is generated by "create_reference_dataset.rmd". Also, only karsten weiss data is loaded here.

```{R Load and Clean DR Dataset}
#this is new karsten/Chan et al 2018 data
  dr_raw <- read_tsv("data/new_karsten_dr_data.txt")

#change colnames, add means and filter missing values 
  #separate df for possibly comparing these values
dr_raw2 <- dr_raw %>%
  rename_all(~c("geneName","comgeneName","hlife_r1","hlife_r2")) %>%
    mutate(hlife = rowMeans(cbind(hlife_r1, hlife_r2), na.rm = TRUE)) %>% 
      filter(is.finite(hlife)) 

#cleaner df
dr_data <- dr_raw2 %>%
          dplyr::select(geneName, comgeneName, hlife)

```


```{R Load Motifs and Codons, echo=TRUE}
#get reference UTRs and CDS sequences from Cheng et al Suppl.
ref_raw <- read_rds("data/Sun_mutation_UTRs.rds")
  #Get coding sequences from ref_raw in a separate vector
  cds_seq <- ref_raw$CDS_seq
  
#Load a motifs list into a vector
motifs_raw <- scan("data/list_motifs.txt", character(), quiet = TRUE)

##Loading of codons_raw is commented out because it is now generated in-code
  #codons_raw <- scan("data/codons.txt", character())

#create combinations of codons
codons_raw <- as_tibble(expand.grid(c("A","T","G","C"), 
                                    c("A","T","G","C"), 
                                    c("A","T","G","C"))) %>%
  #combine 3 columns into a single one
  unite(Var, Var1, Var2, Var3, sep = "") %>% 
    #remove three stop codons and convert to a vector
    filter(!Var %in% c("TAG", "TAA", "TGA")) %>% .$Var 

#load list of 12 Chimeric Terminator Genes  
list_genes <- scan("data/genes_terminators_list.txt", character())

```

```{R Load Ref Dataset}

#Load the file
ref_data <- read_csv("data/ref_data.csv")

#create a subset of motifs and columns from all motifs  
motifs_fil <- motifs_raw[motifs_raw %in% colnames(ref_data)]
codons_inc <- codons_raw[codons_raw %in% colnames(ref_data)]


```

---

### Creating Model Datasets

Decay Data and Motifs/codons count are combined. Combined_full contains both motif counts and codon usage, along with all the genes. combined_sub has removed 10 genes which is used for modelling. combined_sel is used for subseting to only the 10 genes upon which, the model is validated.   

```{R Create model datasets}

model_combined_full <- left_join(ref_data, dr_data, by = "geneName") %>% 
    filter(is.finite(hlife)) %>% 
      #log transform hlife and remove original
      mutate(log2_hlife = log2(hlife)) %>% select(-hlife) %>%
        #reorder columns
        select(geneName, comgeneName, log2_hlife, everything())

#Create a subset of model by taking out selected genes

  #creates an operator for Not %in%
    '%!in%' <- function(x,y)!('%in%'(x,y))
    
  model_combined_sub <- model_combined_full %>%
    filter(geneName %!in% list_genes)
  
#model df containing codon usage and log2(hlife)
model_codons <- model_combined_sub %>% 
  select(geneName, comgeneName, log2_hlife, codons_inc)    
```

Linear modeling is performed first on all codons, and then on both codon usage + motifs. Then, step selection is used to reduce the motifs down (uses AIC) with lower model as all codons, and upper model as combined so that all codons are included. 


```{R LM}
predictors <- paste(c(codons_inc, motifs_fil), collapse = " + ")

lm_codons_dr <- lm(data = model_combined_sub, 
                   paste("log2_hlife", paste(codons_inc, collapse = " + "), sep = " ~ "))

lm_combined_dr <- lm(data = model_combined_sub, 
                   paste("log2_hlife", predictors, sep = " ~ "))

lm_step <- step(lm_combined_dr, 
                scope = list(upper = lm_combined_dr, lower = lm_codons_dr),
                trace = FALSE)

#Create a simple vector of motifs selected by step_lm
motifs_step_sel <- names(coef(lm_step))[nchar(names(coef(lm_step))) > 3][-1]
```

Finally, the output of lm_step is refined to present motifs and coefficients in a table. 

```{R echo = FALSE, results = 'asis'}
tidy_motifs_step <- lm_step %>%
  broom::tidy() %>%
    #perform FDR correction for p-values
    mutate(p_adj_fdr = p.adjust(p.value, method = "fdr")) %>%
      #filter intercept motifs
      filter(term != "(Intercept)" & nchar(term) > 3) %>%
        select(motif = term, coeff = estimate, everything(), -statistic)

kable(tidy_motifs_step, caption = "Table: Motifs and respective coefficients for motifs, selected in combined model for predicting log2(RNA half life) using Codon usage and 3'UTR motifs. Selection was performed using step() function in R, keeping all the codons")
```


Following code uses the combined stepwise selected model to predict the log2(half-life) of the 10 genes that were excluded from the modelling. Then, a plot is constructed for measured vs predicted half-life to visualise the accuracy of the model (cross-validation), and also to find which of the 10 genes are reasonably predicted so that we can easily select candidate 3'UTRs for experimental validation.  

```{R Prediction}

pred_step <- model_combined_full %>%
  filter(geneName %in% list_genes) %>%
    mutate(pred_hlife = predict(lm_step, .))

pred_step_all <- model_combined %>%
  mutate(pred_hlife = predict(lm_step, 
                              model_combined)) %>%
    select(geneName, comgeneName, log2_hlife, pred_hlife, everything())
  
  
plot_predictions <- pred_step %>%
  ggplot(aes(x = log2_hlife, y = pred_hlife, col = paste(geneName, " (", comgeneName, ")"))) +
    geom_point() + 
    scale_x_continuous(limits = c(0, 5)) + 
    scale_y_continuous(limits = c(0, 5)) + 
    geom_abline() + geom_text(aes(label = comgeneName), size = 3, vjust = 0, nudge_y = 0.1, angle = 45) +
    labs(y = "Predicted log2 (Hlife)", 
         x = "Measured log2 (Hlife)",
         colour = "geneID (geneName)",
         title = "Predicted vs Measured Half-life for 10 genes")

#display the plot
plot_predictions

#this is a cleaner, rearranged df just for tabular comparison
pred_motifs_sel <- pred_step %>% 
  select(geneName, comgeneName, log2_hlife, pred_hlife, motifs_step_sel)
```



#####Making Density Plots 
  
Note: This function requres 3UTR_seq and motif frequencies from ref_raw and model_combined_full respesctively

```{R Searching Location for insertion}

#following function accepts a motif, and return a density plot 
motif_density <- function(m_regex){
temp_tib <- model_combined_full %>% 
  select(geneName, comgeneName) %>%
    left_join(., ref_raw[c("genename", "UTR3_seq")], 
              by = c("geneName" = "genename")) %>%
    mutate(test_motif = str_count(UTR3_seq, m_regex)) %>%
      filter(test_motif > 0)

data.frame(do.call(rbind, lapply(temp_tib$UTR3_seq, 
       function(seq){
         str_locate_all(seq, m_regex)[[1]] / nchar(seq)
         }))) %>% 
  ggplot() + geom_density(aes(x = end))

}

#Run for 

```

---

### End of page